---
output: github_document
---
<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# scMINER

<!-- badges: start -->
<!-- badges: end -->

*scMINER* is an R package for preprocessing, QC, clustering, and *hidden driver analysis* of single-cell RNA-seq data. scMINER enables mutual information-based cell clustering, cell-type-specific gene regulatory network (GRN) reverse engineering and protein activity inference, to identify hidden transcriptional factors (TFs) and signaling factors (SIGs) driving cellular lineage differentiation and tissue specific specification. scMINER software consists of three components: [MICA](https://github.com/jyyulab/MICA) (Mutual Information based Clustering analysis), [SJARACNe](https://github.com/jyyulab/SJARACNe) and [MINIE](https://github.com/jyyulab/scMINER/).

## Installation

You can install the development version of scMINER from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("jyyulab/scMINER")
```

## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
library(scMINER)
## basic example code

```

## Tutorial

Read the [documentation](https://jyyulab.github.io/scMINER/site/) for detailed installation instruction and guided analysis. If you'd like to contribute, please open an issue or a pull request in the [github repository](https://github.com/jyyulab/scMINER/issues).

