---
title: "Mutual Information-based network rewiring by SJARACNe"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{network_inference}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(scMINER)
library(ggplot2)
library(dplyr)
```

# Introduction

[SJARACNe](https://academic.oup.com/bioinformatics/article/35/12/2165/5156064) is a scalable software tool for gene network reverse engineering from big data. As an improved implementation of the ARACNe, SJARACNe achieves a dramatic improvement in computational performance in both time and memory usage and implements new features while preserving the network inference accuracy of the original algorithm.

Similar to MICA, SJARACNe is also a component of scMINER framework, and can work seamlessly with the SparseExpressionSet object. The input for MICA can be easily generated from the SparseExpressionSet object by `generateSJARACNeInput()`, and the output of SJARACNe, the networks, can be effortlessly assessed by `drawNetworkQC()` and directly taken for driver activity estimation by `getActivity_individual()` and `getActivity_inBatch()`.

In this tutorial, we will use PBMC14k dataset for demonstration purpose.

<details>
<summary>**How was the PBMC14K dataset generated?**</summary>
```{r, prepare-input-eset, echo=FALSE}
## Read the raw count matrix
pbmc14k_rawCount <- readInput_table(system.file("extdata/demo_pbmc14k/PBMC14k_rawCount.txt.gz", package = "scMINER"), is.geneBYcell = TRUE)

## Read the true labels
pbmc14k_trueLabel <- read.table(system.file("extdata/demo_pbmc14k/PBMC14k_trueLabel.txt.gz", package = "scMINER"), header = T, row.names = 1, sep = "\t", quote = "", stringsAsFactors = FALSE)

## The true_label much cover all cells in the expression matrix
table(colnames(pbmc14k_rawCount) %in% row.names(pbmc14k_trueLabel))

## Create the sparse eSet object using the true_label
pbmc14k_raw.eset <- createSparseEset(input_matrix = pbmc14k_rawCount, cellData = pbmc14k_trueLabel, featureData = NULL, projectID = "PBMC14k", addMetaData = TRUE)

## Filter he sparse eSet object using the auto mode
pbmc14k_filtered.eset <- filterSparseEset(pbmc14k_raw.eset, filter_mode = "auto", filter_type = "both")

## Normalize he sparse eSet object using the auto mode
pbmc14k_log2cpm.eset <- normalizeSparseEset(pbmc14k_filtered.eset, scale_factor = 1000000, log_base = 2, log_pseudoCount = 1)

## Add the clustering results from MICA
pbmc14k_log2cpm.eset <- addMICAoutput(pbmc14k_log2cpm.eset, mica_output_file = system.file("extdata/demo_pbmc14k/PBMC14k/MICA/clustering_UMAP_euclidean_20_2.05.txt", package = "scMINER"), visual_method = "umap")

## Add cell type annotation results
celltype_map <- c(`1`="CD4TN", `2`="CD4TCM", `3`="CD8TN", `4`="NK", `5`="B", `6`="Monocyte", `7`="CD4Treg")
pbmc14k_log2cpm.eset$cell_type <- as.character(celltype_map[pbmc14k_log2cpm.eset$clusterID])
```
</details>
\

# Generate the SJARACNe input files

The network inference should be performed in a **cluster-** or **cell type-specific** basis. For each group (one cluster or cell type), the `generateSJARACNeInput()` function will create a folder named by the label of this group, and generated the standard input files in it:
  
- a "**`.exp.txt`**" file: a tab-separated genes/transcripts/proteins by cells/samples expression matrix with the first two columns being ID and symbol.
- a "**`TF`**" folder containing a "**`.tf.txt`**" file: a list of significant gene/transcript/protein IDs of TF drivers.
- a "**`SIG`**" folder containing a "**`.sig.txt`**" file: a list of significant gene/transcript/protein IDs of SIG drivers.
- a bash script (**`runSJARACNe.sh`**) to run SJARACNe. Further modification is needed to run it.
- a json file (**`config_cwlexec.json`**) containing parameters to run SJARACNe.

Usually, the ground truth of cell types is not available. Then the cluster labels, or cell type annotations of the clusters, can be used for grouping in network rewiring, since it's expected that cells with same cluster label/annotated cell type are of similar gene expression profiles. To generate from annotated cell types, you can run:

```{r generate-sjaracne-inputs-from-cell-type, eval=FALSE}
generateSJARACNeInput(input_eset = pbmc14k_log2cpm.eset, group_name = "cell_type", sjaracne_dir = "/work-path/PBMC14k/SJARACNe/bycelltype", species_type = "hg", driver_type = "TF_SIG")
```

In this case, we know the true labels of cell type, so we use them to define the groups for network inference.

***IMPORTANT NOTE***: scMINER creates a folder for each group using its label. Any illegal characters in the group labels may cause issues in subsequent analysis. To avoid this, scMINER only accept letters(A-Za-z), numbers(0-9), underscores('_') and periods('.').

```{r generate-sjaracne-inputs-from-true-label, eval=FALSE}
## Columns with any illegal characters can not be used for groupping
generateSJARACNeInput(input_eset = pbmc14k_log2cpm.eset, group_name = "trueLabel", sjaracne_dir = "/work-path/PBMC14k/SJARACNe", species_type = "hg", driver_type = "TF_SIG", downSample_N = NULL)
```


# Run the SJARACNe

By default, the `generateSJARACNeInput()` function also generates a `runSJARACNe.sh` file in the folder of each group. This file much be modified before you can run it:

- **removed unneeded lines**: There are usually 4 lines in this file: the lines starting with "sjaracne lsf" are the command lines to run on IBM LSF cluster, while the lines starting with "sjaracne local" are the command lines runing on a single machine (Linux/OSX). Please select the lines based on your situation and remove the others.
- **-n**: number of bootstrap networks to generate. Default: 100.
- **-pc**: p value threshold to select edges in building consensus network. Default: e-2 for single-cell data, e-3 for meta-cell data, and e-5 for bulk sample data.
Please use "sjaracne lsf -h" or "sjaracne local -h" to check more details of arguments available in SJARACNe.

There is another file, `config_cwlexec.json`, available in the folder. It contains the information (e.g. memory request for each step of SJARACNe run) used for LSF job submission. This file is only needed for LSF runs and the default values works well in most cases. If you are running SJARACNe on a big dataset, you may need to request more memory from it.

In this case, we use LSF to run the SJARACNe:
```{shell run-sjaracne-lsf, engine = 'bash', eval=FALSE}
## let's use B cell as an example
# for TF
sjaracne lsf -e /work-path/PBMC14k/SJARACNe/B/B.8572_1902.exp.txt -g /work-path/PBMC14k/SJARACNe/B/TF/B.835_1902.tf.txt -o /work-path/PBMC14k/SJARACNe/B/TF/bt100_pc001 -n 100 -pc 0.01 -j /work-path/PBMC14k/SJARACNe/B/config_cwlexec.json
# for SIG
sjaracne lsf -e /work-path/PBMC14k/SJARACNe/B/B.8572_1902.exp.txt -g /work-path/PBMC14k/SJARACNe/B/SIG/B.4148_1902.sig.txt -o /work-path/PBMC14k/SJARACNe/B/SIG/bt100_pc001 -n 100 -pc 0.01 -j /work-path/PBMC14k/SJARACNe/B/config_cwlexec.json
```

We created a folder named "**bt100_pc001**" in both TF and SIG folders of each group, to save the networks generated under **100 bootstraps** (`-n 100`) and **0.01 consensus p value** (`-pc 0.01`).

To run SJARACNe on a local machine:
```{shell run-sjaracne-local, engine = 'bash', eval=FALSE}
## let's use B cell as an example
# for TF
sjaracne local -e /work-path/PBMC14k/SJARACNe/B/B.8572_1902.exp.txt -g /work-path/PBMC14k/SJARACNe/B/TF/B.835_1902.tf.txt -o /work-path/PBMC14k/SJARACNe/B/TF/bt100_pc001 -n 100 -pc 0.01
# for SIG
sjaracne local -e /work-path/PBMC14k/SJARACNe/B/B.8572_1902.exp.txt -g /work-path/PBMC14k/SJARACNe/B/SIG/B.4148_1902.sig.txt -o /work-path/PBMC14k/SJARACNe/B/SIG/bt100_pc001 -n 100 -pc 0.01
```


# Assess the quality of networks

## Introduction to the network file by SJARACNe

The core output of SJARACNe is the network file named `consensus_network_ncol_.txt`. It contains 9 columns:

- **source**: ID of the source gene, can be the gene symbol;
- **target**: ID of the target gene, can be the gene symbol;
- **source.symbol**: symbol of the source gene;
- **target.symbol**: symbol of the target gene;
- **MI**: mutual information of source-gene pair;
- **pearson**: Pearson correlation coefficient, [-1,1]
- **pearson**: Spearman correlation coefficient, [-1,1]
- **slope**: slop of the regression line, returned by `stats.linregression()`
- **p.value**: p-value for a hypothesis test whose null hypothesis is that the slope is zero, using Wald Test with t-distribution of the test statistic

```{r check-network-file-format}
network_format <- read.table(system.file("extdata/demo_pbmc14k/PBMC14k/SJARACNe/B/TF/bt100_pc001/consensus_network_ncol_.txt", package = "scMINER"), header = T, sep = "\t", quote = "", stringsAsFactors = F)
head(network_format)
```

## Generate network QC report

There is no simple standards to tell the reliability of networks. Empirically, a network with **50-300** target size is good. scMINER provides a function, `drawNetworkQC()`, to quickly assess the quality of networks in batch.

```{r assess-network-quality, eval=FALSE}
network_stats <- drawNetworkQC(sjaracne_dir = "/work-path/PBMC14K/SJARACNe", generate_html = FALSE) # Set `generate_html = TRUE` to generate html-format QC report for each network file
```

```{r show-network-qc-stats}
## The network QC statistics table is saved seperately, for demonstration purposes.
network_stats <- readRDS(system.file("extdata/demo_pbmc14k/PBMC14k/SJARACNe/network_stats.rds", package = "scMINER"))
head(network_stats)
```
The Mutual Information-based network rewiring by SJARACNe is done!











