---
title: "Tutorial for scMINER"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Tutorial for scMINER}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

*scMINER* is an R package for preprocessing, QC, clustering, and *hidden driver analysis* of single-cell RNA-seq data. scMINER enables mutual information-based cell clustering, cell-type-specific gene regulatory network (GRN) reverse engineering and protein activity inference, to identify hidden transcriptional factors (TFs) and signaling factors (SIGs) driving cellular lineage differentiation and tissue specific specification. 

scMINER software consists of three components: 

- [MICA](https://github.com/jyyulab/MICA) (Mutual Information based Clustering analysis) : perform clustering analysis

- [SJARACNe](https://github.com/jyyulab/SJARACNe) : generate cluster specific networks 

- [MINIE](https://github.com/jyyulab/scMINER/)(Mutual Information-based Network Inference Engine) : identify cell-type-specific hidden drivers


## Installation

Users need to install MICA, SJARACNe and scMINER(MINIE) to run the scMINER analysis. 
Using conda to create a virtual environment is strongly recommended. 


```bash
conda create -n scminer python=3.7.6        # Create a python virtual environment
source activate scminer                     # Activate the virtual environment
pip install MICA                            # Install MICA and its dependencies
pip install SJARACNe                        # Install SJARACNe and its dependencies
conda install -c -r r-base=4.0.3            # Install R
conda install -c conda-forge r-devtools     # Install R devtools 
```

In R, you can install the current version of scMINER from [GitHub](https://github.com/) with:

``` r
devtools::install_github("jyyulab/scMINER")
```

### Example1: scMINER Guided Analysis on 14k PBMCs from 10x Genomics

This tutorial introduce you scMINER's basic analysis using a PBMC dataset with 10 sorted populations of 2k cells per population [Zheng et al., 2017](https://www.nature.com/articles/ncomms14049). We remove CD34+ cells, CD4+ helper T cells, and total CD8+ cytotoxic cells from the dataset because of low sorting purity or a significant overlap with other immune cells based on the sorting strategy, which creates a new dataset with seven known cell types and 14k cells in total. The original dataset is freely available under accession number [SRP073767](https://www.ncbi.nlm.nih.gov/sra?term=SRP073767) and [Zenodo](https://zenodo.org/record/3357167#.ZCKhxHZBxaS).

The purpose of scMINER is to identify cell-type specific hidden drivers from single-cell RNASeq dataset. 

The complete step-by-step demo script for this example can be found here: [Example1_PBMC14k.R](https://github.com/jyyulab/scMINER/tree/dev-sherry/demo_scripts/Example1_PBMC14k.R)

##### Step0: Preparations

Make sure you have the MICA, SJARACNe, scMINER package installed. For this example, R package `openxlsx`, `anndata` are required. 

```{r setup}
library(scMINER)
library(openxlsx)
library(anndata)
```


We have designed a function, `scMINER.network.dir.create()` to handle the working directories so that the user can have a better data organization. This function require user to define the main working directory `project_main_dir` and the project name `project_name`. To prevent a previous project with the same `project_main_dir` and `project_name` from being rewritten, it is strongly suggested that you add a time tag to your project_name.

Create directories and folders to save and organize your analysis results.

```{r}
# Define main working directory and project name
project_main_dir <- './test' # user defined main directory for the project, one main directory could have multiple project folders, distinguished by project name.
current_date <- format(Sys.time(), "%Y-%m-%d") # optional, if user like to add current date to name the project folder
project_name <- sprintf('project_%s',current_date) # project name for the project folders under main directory.
```



##### Data loading and preprocessing

-- Read 10x genomics data

-- Quality control and data filtering

-- Normalization and transformation

##### Perform clustering analysis via MICA

-- Generate MICA input

-- Run MICA from the command line

-- Read MICA output

##### Cell type annotation after clustering

-- Marker gene visualization
  
-- Cell type annotation

##### Network generation via SJARACNe

-- Generate SJARACNe input

-- Run SJARACNe from the command line

##### Identify cell-type-specific hidden drivers via MINIE

-- Calculate activity

-- Driver estimation by differential activity analysis
 





### Example2: scMINER Guided Analysis on WT and KO CD8+ T cell in chronic infection model

TOX is a master transcription factor for CD8+ T cell exhaustion during chronic infection. This tutorial introduce you scMINER's basic analysis using a WT and TOX KO CD8+ T dataset (GSE119940) [Yao et al., Nat Immunol 2019](https://www.nature.com/articles/s41590-019-0403-4).

```{r}
library(scMINER)
```

##### Data loading and preprocessing

-- Read 10x genomics data

-- Quality control and data filtering

-- Normalization and transformation

##### Perform clustering analysis via MICA

-- Generate MICA input

-- Run MICA from the command line

-- Read MICA output

##### Cell type annotation after clustering

-- Marker gene visualization
  
-- Cell type annotation

##### WT and KO cell network generation via SJARACNe

-- Generate SJARACNe input

-- Run SJARACNe from the command line

##### Identify genotype-specific hidden drivers via MINIE

-- Calculate activity

-- Driver estimation by differential activity analysis



## Advanced analysis in scMINER

### scMINER Guided Analysis on activity-based clustering

- Generate bulk network by using MetaCell

- Create membership for each cell based on metacell

- Generate eset object for network construction

- Create Pseudobulk eset

- Generate Network


### scMINER Guided Analysis on functional enrichment

- Pathway enrichment

### scMINER Guided Analysis on network visualization

- Read in network

- The rewiring of a specific TF can be output to CytoScape.

### Interoperability between SparseEset, Seurat, SingleCellExperiment, and anndata

- SC data type Conversion

- SparseEset to Seurat

- Seurat to SparseEset




