<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Combine multiple sparse expression set objects — combineSparseEset • scMINER</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Combine multiple sparse expression set objects — combineSparseEset"><meta name="description" content="This function is used to combine the sparse expression set objects. The combined eset object contains all cells and features of all input eset objects. If the eset objects are of
different features, NA values will be generated and automatically imputed by the minimum value of the combined gene expression matrix."><meta property="og:description" content="This function is used to combine the sparse expression set objects. The combined eset object contains all cells and features of all input eset objects. If the eset objects are of
different features, NA values will be generated and automatically imputed by the minimum value of the combined gene expression matrix."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMINER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/quick_tutorial.html">Quick Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="../bookdown/index.html">Full Documentation</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/jyyulab/scMINER"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Combine multiple sparse expression set objects</h1>

      <div class="d-none name"><code>combineSparseEset.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is used to combine the sparse expression set objects. The combined eset object contains all cells and features of all input eset objects. If the eset objects are of
different features, <code>NA</code> values will be generated and automatically imputed by the minimum value of the combined gene expression matrix.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">combineSparseEset</span><span class="op">(</span></span>
<span>  <span class="va">eset_list</span>,</span>
<span>  projectID <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  addPrefix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  addSurfix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  addMetaData <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  imputeNA <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-eset-list">eset_list<a class="anchor" aria-label="anchor" href="#arg-eset-list"></a></dt>
<dd><p>A vector of sparse expression set objects to combine</p></dd>


<dt id="arg-projectid">projectID<a class="anchor" aria-label="anchor" href="#arg-projectid"></a></dt>
<dd><p>A character vector or <code>NULL</code>, set the project names of the eset objects to combine. Default: <code>NULL</code>.</p></dd>


<dt id="arg-addprefix">addPrefix<a class="anchor" aria-label="anchor" href="#arg-addprefix"></a></dt>
<dd><p>A character vector or <code>NULL</code>, add a <strong>prefix</strong> to the cell barcodes of each eset object to combine. It is highly recommended to use a prefix containing letters and/or numbers only, and not starting with numbers. Default: <code>NULL</code>.</p></dd>


<dt id="arg-addsurfix">addSurfix<a class="anchor" aria-label="anchor" href="#arg-addsurfix"></a></dt>
<dd><p>A character vector or <code>NULL</code>, add a <strong>surfix</strong> to the cell barcodes of each eset object to combine. It is highly recommended to use a surfix containing letters and/or numbers only, and not starting with numbers. Default: <code>NULL</code>.</p></dd>


<dt id="arg-addmetadata">addMetaData<a class="anchor" aria-label="anchor" href="#arg-addmetadata"></a></dt>
<dd><p>Logical, whether to update the meta data of cells and features after combination. Default: <code>TRUE</code>.</p></dd>


<dt id="arg-imputena">imputeNA<a class="anchor" aria-label="anchor" href="#arg-imputena"></a></dt>
<dd><p>Logical, whether to impute NA values in combined matrix. If <code>TRUE</code>, the min value of the matrix will be used to replace the NAs. If <code>FALSE</code>, NA values will retain. Default: <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A sparse eset object with combined features and cells of multiple eset objects.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">demo1_mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="readInput_10x.dir.html">readInput_10x.dir</a></span><span class="op">(</span>input_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/demo_inputs/cell_matrix_10x"</span>, package <span class="op">=</span> <span class="st">"scMINER"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                               featureType <span class="op">=</span> <span class="st">"gene_symbol"</span>, removeSuffix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading 10x Genomcis data from: /home/runner/work/_temp/Library/scMINER/extdata/demo_inputs/cell_matrix_10x ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Multiple data modalities were found: Gene Expression, Peaks . Only the gene expression data (under "Gene Expression") was kept.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done! The sparse gene expression matrix has been generated: 500 genes, 100 cells.</span>
<span class="r-in"><span><span class="va">demo1.eset</span> <span class="op">&lt;-</span> <span class="fu"><a href="createSparseEset.html">createSparseEset</a></span><span class="op">(</span>input_matrix <span class="op">=</span> <span class="va">demo1_mtx</span>, projectID <span class="op">=</span> <span class="st">"demo1"</span>, addMetaData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Creating sparse eset from the input_matrix ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Adding meta data based on input_matrix ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done! The sparse eset has been generated: 500 genes, 100 cells.</span>
<span class="r-in"><span><span class="va">demo2_mtx</span> <span class="op">&lt;-</span> <span class="fu"><a href="readInput_table.html">readInput_table</a></span><span class="op">(</span>table_file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/demo_inputs/table_file/demoData2.txt.gz"</span>, package <span class="op">=</span> <span class="st">"scMINER"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                             is.geneBYcell <span class="op">=</span> <span class="cn">TRUE</span>, removeSuffix <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading table file: /home/runner/work/_temp/Library/scMINER/extdata/demo_inputs/table_file/demoData2.txt.gz ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Suffix removal was specified but skipped, since some barcodes do not carry "-1" suffix.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done! The sparse gene expression matrix has been generated: 1000 genes, 100 cells.</span>
<span class="r-in"><span><span class="va">demo2.eset</span> <span class="op">&lt;-</span> <span class="fu"><a href="createSparseEset.html">createSparseEset</a></span><span class="op">(</span>input_matrix <span class="op">=</span> <span class="va">demo2_mtx</span>, projectID <span class="op">=</span> <span class="st">"demo2"</span>, addMetaData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Creating sparse eset from the input_matrix ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Adding meta data based on input_matrix ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done! The sparse eset has been generated: 1000 genes, 100 cells.</span>
<span class="r-in"><span><span class="va">combined.eset</span> <span class="op">&lt;-</span> <span class="fu">combineSparseEset</span><span class="op">(</span>eset_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">demo1.eset</span>, <span class="va">demo2.eset</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                   projectID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                   addPrefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"demo1"</span>, <span class="st">"demo2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                   addSurfix <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                                   addMetaData <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                                   imputeNA <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Combining the input sparse eSets ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NA values were found in the merged matrix and have been replaced by the minimum value:  0 .</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Adding meta data based on merged data matrix ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done! The combined sparse eset has been generated: 1500 genes, 200 cells.</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qingfei Pan, Jiayu Zhou, Liang Ding, Hao Shi, Xinran Dong, Yu Lab @ St Jude Children’s Research Hospital.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

