<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • scMINER</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMINER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/quick_tutorial.html">Quick Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="../bookdown/index.html">Full Documentation</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/jyyulab/scMINER"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="data-preprocessing">Data preprocessing<a class="anchor" aria-label="anchor" href="#data-preprocessing"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="createProjectSpace.html">createProjectSpace()</a></code>

        </dt>
        <dd>Create a project space for scMINER analysis</dd>
      </dl></div><div class="section level2">

      <h3 id="data-intake">Data intake<a class="anchor" aria-label="anchor" href="#data-intake"></a></h3>
      <div class="section-desc"><p>Functions to read input data from multiple formats</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="readInput_10x.dir.html">readInput_10x.dir()</a></code>

        </dt>
        <dd>Read the input data generated by 10x Genomics from a directory</dd>
      </dl><dl><dt>

          <code><a href="readInput_10x.h5.html">readInput_10x.h5()</a></code>

        </dt>
        <dd>Read the input data generated by 10x Genomics from the HDF5 file</dd>
      </dl><dl><dt>

          <code><a href="readInput_h5ad.html">readInput_h5ad()</a></code>

        </dt>
        <dd>Read the h5ad file</dd>
      </dl><dl><dt>

          <code><a href="readInput_table.html">readInput_table()</a></code>

        </dt>
        <dd>Read the table format file</dd>
      </dl></div><div class="section level2">

      <h3 id="sparseeset-manipulation">SparseEset manipulation<a class="anchor" aria-label="anchor" href="#sparseeset-manipulation"></a></h3>
      <div class="section-desc"><p>Functions that help create and filter SparseEset objects</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="createSparseEset.html">createSparseEset()</a></code>

        </dt>
        <dd>Create a sparse expression set object from a data matrix</dd>
      </dl><dl><dt>

          <code><a href="combineSparseEset.html">combineSparseEset()</a></code>

        </dt>
        <dd>Combine multiple sparse expression set objects</dd>
      </dl><dl><dt>

          <code><a href="updateSparseEset.html">updateSparseEset()</a></code>

        </dt>
        <dd>Update the slots and/or meta data of the sparse eset object</dd>
      </dl><dl><dt>

          <code><a href="filterSparseEset.html">filterSparseEset()</a></code>

        </dt>
        <dd>Filter the cells and/or features of sparse eset object using automatic or self-customized cutoffs</dd>
      </dl><dl><dt>

          <code><a href="normalizeSparseEset.html">normalizeSparseEset()</a></code>

        </dt>
        <dd>Normalize and log-transform the sparse eset object</dd>
      </dl><dl><dt>

          <code><a href="drawSparseEsetQC.html">drawSparseEsetQC()</a></code>

        </dt>
        <dd>Generate a quality control report from sparse eset object</dd>
      </dl></div><div class="section level2">
      <h2 id="mi-based-clustering-analysis">MI-based clustering analysis<a class="anchor" aria-label="anchor" href="#mi-based-clustering-analysis"></a></h2>

      <div class="section-desc"><p>Functions that mainly used in clustering analysis</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="generateMICAinput.html">generateMICAinput()</a></code>

        </dt>
        <dd>Generate the standard input files for MICA from sparse eset object</dd>
      </dl><dl><dt>

          <code><a href="addMICAoutput.html">addMICAoutput()</a></code>

        </dt>
        <dd>Add the MICA output (cluster labels and UMAP/tSNE coordinates) to sparse eset object</dd>
      </dl><dl><dt>

          <code><a href="MICAplot.html">MICAplot()</a></code>

        </dt>
        <dd>Draw a scatter plot showing the coordinates and cluster id of each cell</dd>
      </dl></div><div class="section level2">
      <h2 id="mi-based-network-inference">MI-based network inference<a class="anchor" aria-label="anchor" href="#mi-based-network-inference"></a></h2>

      <div class="section-desc"><p>Functions that principally operate for network inference</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getDriverList.html">getDriverList()</a></code>

        </dt>
        <dd>Extract the pre-defined driver lists of human or mouse</dd>
      </dl><dl><dt>

          <code><a href="generateSJARACNeInput.html">generateSJARACNeInput()</a></code>

        </dt>
        <dd>Generate the standard input files for SJARACNe from sparse eset object</dd>
      </dl><dl><dt>

          <code><a href="drawNetworkQC.html">drawNetworkQC()</a></code>

        </dt>
        <dd>Assess the quality of each network generated by SJARACNe</dd>
      </dl><dl><dt>

          <code><a href="getActivity_individual.html">getActivity_individual()</a></code>

        </dt>
        <dd>Calculate driver activities per group from network files</dd>
      </dl><dl><dt>

          <code><a href="getActivity_inBatch.html">getActivity_inBatch()</a></code>

        </dt>
        <dd>Calculate driver activities in batch from the SJARACNe directory</dd>
      </dl></div><div class="section level2">
      <h2 id="data-visualization-and-sharing">Data visualization and sharing<a class="anchor" aria-label="anchor" href="#data-visualization-and-sharing"></a></h2>

      <div class="section-desc"><p>Functions that visulize the data analysis and prepare inputs for scMINER Portal</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="feature_vlnplot.html">feature_vlnplot()</a></code>

        </dt>
        <dd>Violin plot showing the expression or activity of selected features by self-defined groups</dd>
      </dl><dl><dt>

          <code><a href="feature_boxplot.html">feature_boxplot()</a></code>

        </dt>
        <dd>Box plot showing the expression or activity of selected features by self-defined groups</dd>
      </dl><dl><dt>

          <code><a href="feature_scatterplot.html">feature_scatterplot()</a></code>

        </dt>
        <dd>Scatter plot showing the expression or activity of selected features on UMAP or t-SNE coordinates</dd>
      </dl><dl><dt>

          <code><a href="feature_bubbleplot.html">feature_bubbleplot()</a></code>

        </dt>
        <dd>Bubble blot showing the expression or activity of selected features by self-defined groups</dd>
      </dl><dl><dt>

          <code><a href="feature_heatmap.html">feature_heatmap()</a></code>

        </dt>
        <dd>Heatmap showing the expression or activity of selected features by self-defined groups</dd>
      </dl><dl><dt>

          <code><a href="draw_barplot.html">draw_barplot()</a></code>

        </dt>
        <dd>Bar plot showing the cell composition of self-defined groups</dd>
      </dl><dl><dt>

          <code><a href="draw_bubbleplot.html">draw_bubbleplot()</a></code>

        </dt>
        <dd>Bubble plot showing the signature scores by self-defined groups</dd>
      </dl><dl><dt>

          <code><a href="generatePortalInputs.html">generatePortalInputs()</a></code>

        </dt>
        <dd>Prepare the standard input files for scMINER Portal</dd>
      </dl></div><div class="section level2">
      <h2 id="differential-analysis">Differential analysis<a class="anchor" aria-label="anchor" href="#differential-analysis"></a></h2>

      <div class="section-desc"><p>Functions that identify the differntially expressed genes (DEGs) or differentially activated drivers (DADs)</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="getDE.html">getDE()</a></code>

        </dt>
        <dd>Perform differential expression analysis on expression set object</dd>
      </dl><dl><dt>

          <code><a href="getDA.html">getDA()</a></code>

        </dt>
        <dd>Perform differential activity analysis on expression set</dd>
      </dl><dl><dt>

          <code><a href="getTopFeatures.html">getTopFeatures()</a></code>

        </dt>
        <dd>Pick the top genes/drivers for each group from differential analysis results</dd>
      </dl></div><div class="section level2">
      <h2 id="helper-functions">Helper functions<a class="anchor" aria-label="anchor" href="#helper-functions"></a></h2>

      <div class="section-desc"><p>Functions that make the primary functions simpler</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="combinePvalVector.html">combinePvalVector()</a></code>

        </dt>
        <dd>Combine P values using Fisher's method or Stouffer's method</dd>
      </dl><dl><dt>

          <code><a href="compare2groups.html">compare2groups()</a></code>

        </dt>
        <dd>Perform differential analysis between two groups</dd>
      </dl><dl><dt>

          <code><a href="z_normalization.html">z_normalization()</a></code>

        </dt>
        <dd>Scale a numeric vector using Z-normalization</dd>
      </dl><dl><dt>

          <code><a href="get_net2target_list.html">get_net2target_list()</a></code>

        </dt>
        <dd>Convert a txt network file to a list</dd>
      </dl><dl><dt>

          <code><a href="get_target_list2matrix.html">get_target_list2matrix()</a></code>

        </dt>
        <dd>Convert ther list generated by <code><a href="../reference/get_net2target_list.html">get_net2target_list()</a></code> to a matrix of signed mutual information</dd>
      </dl><dl><dt>

          <code><a href="cal_Activity.html">cal_Activity()</a></code>

        </dt>
        <dd>Calculate driver activities from gene expression matrix and networks</dd>
      </dl><dl><dt>

          <code><a href="SparseExpressionSet-class.html">SparseExpressionSet-class</a></code>

        </dt>
        <dd>SparseExpressionSet</dd>
      </dl><dl><dt>

          <code><a href="pbmc14k_rawCount.html">pbmc14k_rawCount</a></code>

        </dt>
        <dd>Raw count matrix of PBMC14k dataset</dd>
      </dl><dl><dt>

          <code><a href="pbmc14k_expression.eset.html">pbmc14k_expression.eset</a></code>

        </dt>
        <dd>SparseEset object of PBMC14k dataset</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qingfei Pan, Jiayu Zhou, Liang Ding, Hao Shi, Xinran Dong, Yu Lab @ St Jude Children’s Research Hospital.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

