<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Read the input data generated by 10x Genomics from the HDF5 file — readInput_10x.h5 • scMINER</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Read the input data generated by 10x Genomics from the HDF5 file — readInput_10x.h5"><meta name="description" content="This function is used to read the gene expression data from the
HDF5 file generated by CellRanger pipeline of 10x Genomics. This function can
automatically distinguish the data of different modalities (e.g. expression
data, ATAC data) and retains the gene expression data only. The **hdf5r**
package is needed to use this function."><meta property="og:description" content="This function is used to read the gene expression data from the
HDF5 file generated by CellRanger pipeline of 10x Genomics. This function can
automatically distinguish the data of different modalities (e.g. expression
data, ATAC data) and retains the gene expression data only. The **hdf5r**
package is needed to use this function."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMINER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/quick_tutorial.html">Quick Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="../bookdown/index.html">Full Documentation</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/jyyulab/scMINER"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Read the input data generated by 10x Genomics from the HDF5 file</h1>

      <div class="d-none name"><code>readInput_10x.h5.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is used to read the gene expression data from the
HDF5 file generated by CellRanger pipeline of 10x Genomics. This function can
automatically distinguish the data of different modalities (e.g. expression
data, ATAC data) and retains the gene expression data only. The <code>**hdf5r**</code>
package is needed to use this function.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">readInput_10x.h5</span><span class="op">(</span></span>
<span>  <span class="va">h5_file</span>,</span>
<span>  featureType <span class="op">=</span> <span class="st">"gene_symbol"</span>,</span>
<span>  removeSuffix <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  addPrefix <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-h-file">h5_file<a class="anchor" aria-label="anchor" href="#arg-h-file"></a></dt>
<dd><p>H5 file generated by CellRanger pipeline of 10x Genomics</p></dd>


<dt id="arg-featuretype">featureType<a class="anchor" aria-label="anchor" href="#arg-featuretype"></a></dt>
<dd><p>Character, feature type to use as the gene name of
expression matrix: <code>"gene_symbol"</code> (the default) or <code>"gene_id"</code>.</p></dd>


<dt id="arg-removesuffix">removeSuffix<a class="anchor" aria-label="anchor" href="#arg-removesuffix"></a></dt>
<dd><p>Logical, whether to remove the suffix "<strong>-1</strong>" when
present in all cell barcodes. Default: <code>TRUE</code>.</p></dd>


<dt id="arg-addprefix">addPrefix<a class="anchor" aria-label="anchor" href="#arg-addprefix"></a></dt>
<dd><p>Character or <code>NULL</code>, add a prefix to the cell barcodes, like
Sample ID. It is highly recommended to use a prefix containing letters
and/or numbers only, and not starting with numbers. Default: <code>NULL</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A sparse gene expression matrix of raw UMI counts, genes by cells</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">h5_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/demo_inputs/hdf5_10x/demoData3.h5"</span>, package <span class="op">=</span> <span class="st">"scMINER"</span><span class="op">)</span> <span class="co"># path to hdf5 file</span></span></span>
<span class="r-in"><span><span class="va">sparseMatrix</span> <span class="op">&lt;-</span> <span class="fu">readInput_10x.h5</span><span class="op">(</span><span class="va">h5_file</span>,</span></span>
<span class="r-in"><span>                                 featureType <span class="op">=</span> <span class="st">"gene_symbol"</span>,</span></span>
<span class="r-in"><span>                                 removeSuffix <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                                 addPrefix <span class="op">=</span> <span class="st">"demoSample"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading 10x Genomics data from: /home/runner/work/_temp/Library/scMINER/extdata/demo_inputs/hdf5_10x/demoData3.h5 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Checking HDF5 file format ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Format check passed!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done! The sparse gene expression matrix has been generated: 36601 genes, 182 cells.</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qingfei Pan, Jiayu Zhou, Liang Ding, Hao Shi, Xinran Dong, Yu Lab @ St Jude Children’s Research Hospital.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

