<!DOCTYPE html>

<html lang="en-us">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=Edge">
  

  <title>Sample Analysis with PBMC(12k) scRNA-seq data - </title>
  <link rel="stylesheet" href="http://localhost:4000/assets/css/just-the-docs.css">

  

  
  <script type="text/javascript" src="http://localhost:4000/assets/js/just-the-docs.js"></script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
</head>

<body>

  <div class="page-wrap">
    <div class="side-bar">
      <a href="http://localhost:4000" class="site-title fs-6 lh-tight"></a>
      <span class="fs-3"><button class="js-main-nav-trigger navigation-list-toggle btn btn-outline" type="button" data-text-toggle="Hide">Menu</button></span>
      <div class="navigation main-nav js-main-nav">
        <nav role="navigation" aria-label="Main navigation">
  <ul class="navigation-list">
    
    
      
        
          <li class="navigation-list-item active">
            
            <a href="http://localhost:4000/docs/index-test.html" class="navigation-list-link"></a>
            
          </li>
        
      
    
      
        
          <li class="navigation-list-item">
            
            <a href="http://localhost:4000/" class="navigation-list-link">Welcome to scMINER</a>
            
          </li>
        
      
    
      
        
          <li class="navigation-list-item">
            
            <a href="http://localhost:4000/docs/MICA.html" class="navigation-list-link">Run MICA locally</a>
            
          </li>
        
      
    
      
        
          <li class="navigation-list-item">
            
            <a href="http://localhost:4000/docs/MICA_LSF.html" class="navigation-list-link">Run MICA on LSF</a>
            
          </li>
        
      
    
      
        
          <li class="navigation-list-item active">
            
            <a href="http://localhost:4000/docs/PBMC-12k.html" class="navigation-list-link active">Sample Analysis with PBMC(12k) scRNA-seq data</a>
            
          </li>
        
      
    
      
        
          <li class="navigation-list-item">
            
            <a href="http://localhost:4000/docs/PBMC-12k-network.html" class="navigation-list-link">Cell-type specific network analysis</a>
            
          </li>
        
      
    
  </ul>
</nav>

      </div>
      <footer role="contentinfo" class="site-footer">
        <p class="text-small text-grey-dk-000 mb-0">This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</p>
      </footer>
    </div>
    <div class="main-content-wrap js-main-content" tabindex="0">
      <div class="page-header">
        <div class="main-content">
          
          
        </div>
      </div>
      <div class="main-content">
        
          
        
        <div id="main-content" class="page-content" role="main">
          <h1 class="no_toc" id="analysis-on-pbmc12k-scrna-seq-data-via-scminer">Analysis on PBMC(12k) scRNA-seq data via scMINER</h1>
<p>Here we demonstrate our pipeline using PBMC (10x genmomics) scRNA-seq data [link to data matrix]. Full data contains 68k cells(link to 10x website), in order to provide a quicker guidance, we’ve down sampled this data to 12k cells.
Original data website can be downloaded <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/fresh_68k_pbmc_donor_a">here</a>.</p>

<h2 class="no_toc text-delta" id="table-of-contents">Table of contents</h2>

<ol id="markdown-toc">
  <li><a href="#installation" id="markdown-toc-installation">Installation</a></li>
  <li><a href="#scrna-seq-data-preprocessing" id="markdown-toc-scrna-seq-data-preprocessing">scRNA-seq data preprocessing</a>    <ol>
      <li><a href="#read-10x-genomics-data" id="markdown-toc-read-10x-genomics-data">Read 10x genomics data</a></li>
      <li><a href="#down-sampling" id="markdown-toc-down-sampling">Down sampling</a></li>
      <li><a href="#quality-control" id="markdown-toc-quality-control">Quality control</a></li>
      <li><a href="#normalization-and-transformation" id="markdown-toc-normalization-and-transformation">Normalization and transformation</a></li>
      <li><a href="#generate-mica-input-and-command" id="markdown-toc-generate-mica-input-and-command">Generate MICA input and command</a></li>
    </ol>
  </li>
  <li><a href="#run-mica-clustering" id="markdown-toc-run-mica-clustering">Run MICA clustering</a></li>
  <li><a href="#cell-type-analysis-from-mica-output" id="markdown-toc-cell-type-analysis-from-mica-output">Cell type analysis from MICA output</a>    <ol>
      <li><a href="#reading-mica-output" id="markdown-toc-reading-mica-output">Reading MICA output</a></li>
      <li><a href="#marker-gene-highlighting" id="markdown-toc-marker-gene-highlighting">Marker gene highlighting</a></li>
      <li><a href="#assign-cell-type-to-cluster" id="markdown-toc-assign-cell-type-to-cluster">Assign cell type to cluster</a></li>
    </ol>
  </li>
  <li><a href="#network-generation-via-sjaracne" id="markdown-toc-network-generation-via-sjaracne">Network generation via SJARACNe</a></li>
  <li><a href="#find-cell-type-specific-master-regulator" id="markdown-toc-find-cell-type-specific-master-regulator">Find cell type specific master regulator</a>    <ol>
      <li><a href="#calculate-inferred-activity" id="markdown-toc-calculate-inferred-activity">Calculate Inferred activity</a></li>
      <li><a href="#find-differential-activity-tf" id="markdown-toc-find-differential-activity-tf">Find Differential activity TF</a></li>
    </ol>
  </li>
</ol>

<hr />
<h2 id="installation">Installation</h2>

<p>scMINER is based on python and R, with a single R package glued all essential functions together.</p>

<h2 id="scrna-seq-data-preprocessing">scRNA-seq data preprocessing</h2>
<p>This can be done by any scRNA-seq preprocessing pipeline. We encourage user to feed in all genes from your data for MICA, instead of highly variable genes only. Here, in order to stick to the focus, we only demonstrate a quick function to conduct gene/cell filtering, without any data exploratory visualization.</p>

<h3 id="read-10x-genomics-data">Read 10x genomics data</h3>
<p>Read 10x genomics data wih function embeded in scMINER package.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readscRNAseqData</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"PBMC68k_input/"</span><span class="p">,</span><span class="n">is.10x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="down-sampling">Down sampling</h3>
<p>In order to provide a quicker guidance, we’ve down sampled this data to 12k cells.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">1</span><span class="p">)</span><span class="w">
</span><span class="n">d</span><span class="m">.12</span><span class="n">k</span><span class="o">&lt;-</span><span class="n">d</span><span class="p">[,</span><span class="n">sample</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">d</span><span class="p">),</span><span class="m">12000</span><span class="p">)]</span><span class="w">
</span><span class="nf">dim</span><span class="p">(</span><span class="n">d</span><span class="m">.12</span><span class="n">k</span><span class="p">)</span><span class="w"> </span><span class="c1">#[1] 32738 12000</span><span class="w">
</span></code></pre></div></div>

<h3 id="quality-control">Quality control</h3>
<p>Quality control includes two function with different purpose, 1. <code class="highlighter-rouge">preMICA.QA</code>, which can generate visualizations to help assist scRNA-seq data quality. 2. <code class="highlighter-rouge">preMICA.QC</code>, which uses pre-defined threshold (either calculated from <code class="highlighter-rouge">preMICA.QA</code> or via manual input). We recommend first run <code class="highlighter-rouge">preMICA.QA</code> then <code class="highlighter-rouge">preMICA.QC</code>. Both of these two functions have versions with or without rmd output. The rmd output will include all computed threshold an d visualizations.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">d</span><span class="o">&lt;-</span><span class="n">preMICA.QA</span><span class="p">(</span><span class="n">raw.data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">d</span><span class="m">.12</span><span class="n">k</span><span class="p">,</span><span class="w">
              </span><span class="n">feature.data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">d</span><span class="m">.68</span><span class="n">k</span><span class="o">$</span><span class="n">feature.data</span><span class="p">,</span><span class="w">
              </span><span class="n">project.name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"PBMC12k_sample"</span><span class="p">,</span><span class="w">
              </span><span class="n">output_rmd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w">
              </span><span class="n">plot.dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"./QC/"</span><span class="p">,</span><span class="w">
              </span><span class="n">cell_percentage</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.005</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>
<p>The output will be stored as a list with data matrix, meta data, feature data as well as the plot directory and calculated cutoffs(under <code class="highlighter-rouge">cal.cutoffs</code> slot.</p>

<p>If you set <code class="highlighter-rouge">plotting = TRUE</code>, then three visualization plot will be generated for gene/cell quality assessment.</p>

<p>One is the total number of cell expressed by each gene</p>

<p><img src="./plots/1_1_Gene_QCmetrics_before_filtering.png" alt="drawing" width="550" /></p>

<p>Second plot will visualize total UMI count, and total number of gene expressed.
<img src="./plots/1_2_Cell_QC_1.png" alt="" /></p>

<p>Third plot will visualize mitochondria percentage, and spike-in percentage for each cell.
<img src="./plots/1_3_Cell_QC_2.png" alt="" /></p>

<p>Then you could run quality control as simple as</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">d.sel</span><span class="o">&lt;-</span><span class="n">preMICA.QC</span><span class="p">(</span><span class="n">d</span><span class="p">,</span><span class="n">project.name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"PBMC12k_sample"</span><span class="p">,</span><span class="w"> </span><span class="n">output_rmd</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>
<p>If <code class="highlighter-rouge">output_rmd=TRUE</code>,then visualization and corresponding stats will be reported in a html file.</p>

<h3 id="normalization-and-transformation">Normalization and transformation</h3>
<p>In MICA, we don’t provide a vriety of methods to do normalizaton, you can use your own prefered normalization method. Here we provide an example of doing CPM and log2 transformation.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">norm</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10e6</span><span class="w">
</span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sweep</span><span class="p">(</span><span class="n">d.sel</span><span class="o">$</span><span class="n">filtered.data</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">norm</span><span class="o">/</span><span class="n">unname</span><span class="p">(</span><span class="n">d.sel</span><span class="o">$</span><span class="n">meta.data</span><span class="o">$</span><span class="n">nUMI.total</span><span class="p">),</span><span class="w"> </span><span class="s1">'*'</span><span class="p">)</span><span class="w">

</span><span class="c1"># log transformation</span><span class="w">
</span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="n">data</span><span class="m">+1</span><span class="p">,</span><span class="n">base</span><span class="o">=</span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="n">d.sel</span><span class="o">$</span><span class="n">data</span><span class="o">&lt;-</span><span class="n">data</span><span class="w">

</span></code></pre></div></div>

<h3 id="generate-mica-input-and-command">Generate MICA input and command</h3>
<p>After reviewing all visualization and conduct filtering, you can go ahead and generate clustering(MICA) input, which will be a cell by gene txt file. If your data was not generated by 10x genomics, you can prepare your own filtering and use this step for clustering Input generation, please note that <code class="highlighter-rouge">you should always feed MICA the log or log2 transformed data</code>.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">generateMICAinput</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">d</span><span class="p">,</span><span class="n">filename</span><span class="o">=</span><span class="s2">"PBMC12k_MICA_input.txt"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>This will output a txt file containing filtered expression matrix for MICA. 
We also provide a function for you to generate MICA command without writing your own scripts. If you set <code class="highlighter-rouge">host=lsf</code>, then you need to define <code class="highlighter-rouge">queue</code> required, and <code class="highlighter-rouge">memory</code> (optional). In <code class="highlighter-rouge">num_cluster</code>, you can input a vector of number of K to achieve clustering membership for different k simultaneously.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">generate_MICA_cmd</span><span class="p">(</span><span class="n">save_sh_at</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"./PBMC12k_v3/"</span><span class="p">,</span><span class="w">
                  </span><span class="n">input_file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"./PBMC12k_v3/PBMC12k_MICA_input.txt"</span><span class="p">,</span><span class="w">
                  </span><span class="n">project_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"PBMC12k"</span><span class="p">,</span><span class="n">num_cluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">8</span><span class="p">,</span><span class="m">9</span><span class="p">,</span><span class="m">10</span><span class="p">,</span><span class="m">12</span><span class="p">,</span><span class="m">13</span><span class="p">,</span><span class="m">14</span><span class="p">,</span><span class="m">15</span><span class="p">),</span><span class="w">
                  </span><span class="n">host</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lsf"</span><span class="p">,</span><span class="w"> </span><span class="n">queue</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="p">[</span><span class="n">your_queue_name</span><span class="p">],</span><span class="w">
                  </span><span class="n">output_path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"./"</span><span class="p">,</span><span class="n">queue</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"standard"</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>

<h2 id="run-mica-clustering">Run MICA clustering</h2>

<p>MICA is implemented in Python. If you would like to install MICA, please infer <a href="https://github.com/jyyulab/MICA">MICA github page</a>.There are several parameters for you to choose when running MICA. A more comprehensive tutorial could be found <a href="./MICA.md">here for local</a> and <a href="./MICA_LSF.md">here for lsf</a>.Here recommend save your working directory before running MICA.</p>

<h2 class="d-inline-block" id="cell-type-analysis-from-mica-output">Cell type analysis from MICA output</h2>

<p>First, after clustering via MICA(see [MICA] (/docs/MICA.html), you can load MICA output (in .txt) as well as input expression matrix in R under an <code class="highlighter-rouge">expressionSet</code>. This is going to be the major data structure we used for downstream analysis in R.</p>

<h3 id="reading-mica-output">Reading MICA output</h3>
<p>You can start with one MICA membership and study your optimal number of cluster with cell type specific markers.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">eset.12k</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readMICAoutput</span><span class="p">(</span><span class="n">Obj</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">d.sel</span><span class="p">,</span><span class="n">load_clust_label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> 
</span><span class="n">output_file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"MICA/PBMC12k_k8_tsne_ClusterMem.txt"</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>

<p>To visualize MICA label or other metadata, you can use function <code class="highlighter-rouge">MICAplot</code>.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">MICAplot</span><span class="p">(</span><span class="n">input_eset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eset.12k</span><span class="p">,</span><span class="n">visualize</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">'tSNE'</span><span class="p">,</span><span class="n">X</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"X"</span><span class="p">,</span><span class="n">Y</span><span class="o">=</span><span class="s2">"Y"</span><span class="p">,</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"label"</span><span class="p">,</span><span class="n">pct</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>
<p><img src="./plots/3_0_MICA_k8.png" alt="" /></p>

<h3 id="marker-gene-highlighting">Marker gene highlighting</h3>

<p>Picked marker genes could be visualized on t-SNE scatterplot, heatmap or violinplot. This will help pick up a reasonable number of cluster.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gn.sel</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s2">"CD3D"</span><span class="p">,</span><span class="s2">"CD27"</span><span class="p">,</span><span class="s2">"IL7R"</span><span class="p">,</span><span class="s2">"SELL"</span><span class="p">,</span><span class="s2">"CCR7"</span><span class="p">,</span><span class="s2">"IL32"</span><span class="p">,</span><span class="s2">"GZMA"</span><span class="p">,</span><span class="w">
          </span><span class="s2">"GZMK"</span><span class="p">,</span><span class="s2">"DUSP2"</span><span class="p">,</span><span class="s2">"CD8A"</span><span class="p">,</span><span class="s2">"GZMH"</span><span class="p">,</span><span class="s2">"GZMB"</span><span class="p">,</span><span class="s2">"CD79A"</span><span class="p">,</span><span class="s2">"CD79B"</span><span class="p">,</span><span class="s2">"CD86"</span><span class="p">,</span><span class="s2">"CD14"</span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">feature_highlighting</span><span class="p">(</span><span class="n">input_eset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eset.12k</span><span class="p">,</span><span class="n">target</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gn.sel</span><span class="p">,</span><span class="n">ylabel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"log2Exp"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="o">=</span><span class="s2">"X"</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="s2">"Y"</span><span class="p">,</span><span class="n">title.size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>
<p><img src="./plots/3_1_gene_highlighting.png" alt="" /></p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">feature_vlnplot</span><span class="p">(</span><span class="n">eset.12k</span><span class="p">,</span><span class="n">target</span><span class="o">=</span><span class="n">gn.sel</span><span class="p">,</span><span class="n">feature</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"geneSymbol"</span><span class="p">,</span><span class="w">
</span><span class="n">group_tag</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"label"</span><span class="p">,</span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="n">ylabel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"log2Exp"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p><img src="./plots/3_2_gene_highlighting_vlnplot.png" alt="" /></p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">feature_heatmap</span><span class="p">(</span><span class="n">eset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">eset.12k</span><span class="p">,</span><span class="n">target</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gn.sel</span><span class="p">,</span><span class="n">group_tag</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"label"</span><span class="p">,</span><span class="w">
			 </span><span class="n">save_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w">
             </span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"log2_expression"</span><span class="p">,</span><span class="n">plot_name</span><span class="o">=</span><span class="s2">"./GeneHeatmap.png"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="./plots/3_3_Marker_heatmaps.png" alt="" /></p>

<h3 id="assign-cell-type-to-cluster">Assign cell type to cluster</h3>

<p>Here we curated a reference signature list of 8 immune cell types(link) for cell type annotation. In <code class="highlighter-rouge">AssignCellTypes.bbp</code> function, we calculated cell type scores for each clusters, and visualize scores using heatmap.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ref</span><span class="o">&lt;-</span><span class="n">read.xlsx</span><span class="p">(</span><span class="s2">"Immune_signatures.xlsx"</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span><span class="w">
</span><span class="o">&gt;</span><span class="w"> </span><span class="n">head</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span><span class="w">
  </span><span class="n">celltype</span><span class="w"> </span><span class="n">markers</span><span class="w"> </span><span class="n">weight</span><span class="w">
</span><span class="m">1</span><span class="w">   </span><span class="n">NaiveT</span><span class="w">    </span><span class="n">SELL</span><span class="w">      </span><span class="m">1</span><span class="w">
</span><span class="m">2</span><span class="w">   </span><span class="n">NaiveT</span><span class="w">    </span><span class="n">CCR7</span><span class="w">      </span><span class="m">1</span><span class="w">
</span><span class="m">3</span><span class="w">     </span><span class="n">Tmem</span><span class="w">    </span><span class="n">IL7R</span><span class="w">      </span><span class="m">1</span><span class="w">
</span><span class="m">4</span><span class="w">     </span><span class="n">Tmem</span><span class="w">    </span><span class="n">CD27</span><span class="w">      </span><span class="m">1</span><span class="w">
</span><span class="m">5</span><span class="w">     </span><span class="n">Tmem</span><span class="w">    </span><span class="n">IL32</span><span class="w">      </span><span class="m">1</span><span class="w">
</span><span class="m">6</span><span class="w">     </span><span class="n">Tmem</span><span class="w">    </span><span class="n">GZMA</span><span class="w">     </span><span class="m">-1</span><span class="w">

</span><span class="n">p</span><span class="o">&lt;-</span><span class="n">AssignCellTypes.bbp</span><span class="p">(</span><span class="n">ref</span><span class="o">=</span><span class="n">ref</span><span class="p">,</span><span class="n">eset</span><span class="o">=</span><span class="n">eset.12k</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>
<p><img src="./plots/3_4_MICA_cluster_score.png" alt="" /></p>

<p>We recommend assign your celltype as factors in your expression set.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">indx</span><span class="o">&lt;-</span><span class="n">factor</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"NaiveT"</span><span class="p">,</span><span class="s2">"Tmem"</span><span class="p">,</span><span class="s2">"CD8em"</span><span class="p">,</span><span class="s2">"CD8eff"</span><span class="p">,</span><span class="s2">"NK"</span><span class="p">,</span><span class="s2">"Bcell"</span><span class="p">,</span><span class="s2">"DC"</span><span class="p">,</span><span class="s2">"Mo"</span><span class="p">),</span><span class="w">
				</span><span class="n">levels</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"NaiveT"</span><span class="p">,</span><span class="s2">"Tmem"</span><span class="p">,</span><span class="s2">"CD8em"</span><span class="p">,</span><span class="s2">"CD8eff"</span><span class="p">,</span><span class="s2">"NK"</span><span class="p">,</span><span class="s2">"Bcell"</span><span class="p">,</span><span class="s2">"DC"</span><span class="p">,</span><span class="s2">"Mo"</span><span class="p">))</span><span class="w">
</span><span class="n">eset.12k</span><span class="o">$</span><span class="n">celltype</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">indx</span><span class="p">[</span><span class="n">eset.12k</span><span class="o">$</span><span class="n">label</span><span class="p">]</span><span class="w">
</span></code></pre></div></div>

<h2 id="network-generation-via-sjaracne">Network generation via SJARACNe</h2>

<p>In order to generate cell type/group/cluster specific network, group information should be stored under <code class="highlighter-rouge">pData([your_expressionSet])</code>. And R function <code class="highlighter-rouge">generateSJAracneInput</code> will help to partition your expression matrix and conduct a loose filtering of your scRNA-seq data(filter about 0 expressed genes in cluster). Besides, a reference TF list should be provided as <code class="highlighter-rouge">tf.ref</code> to guide hub gene selection. Each group will create one directory which contains filtered expression matrix in .exp format, as long as the filtered TF list in .txt.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">generateSJAracneInput</span><span class="p">(</span><span class="n">eset</span><span class="o">=</span><span class="n">eset.12k</span><span class="p">,</span><span class="n">tf.ref</span><span class="o">=</span><span class="n">tf.ref</span><span class="p">,</span><span class="n">wd.src</span><span class="o">=</span><span class="s2">"Sjaracne/"</span><span class="p">,</span><span class="w"> </span><span class="n">group_tag</span><span class="o">=</span><span class="s2">"celltype"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><em>Warning:</em> SJARACNe has not been integrated into scMINER yet, please consult <a href="https://github.com/jyyulab/SJARACNe">here</a> for installation and basic usage. Here recommend save your working directory before running SJARACNe.</p>

<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">indir</span><span class="o">=</span>~/PBMC12K/SJARACNE_PBMC12K/

<span class="k">for </span>i <span class="k">in</span> <span class="k">$(</span><span class="nb">ls</span> <span class="nt">-d</span> <span class="k">*</span>/ | <span class="nb">cut</span> <span class="nt">-f1</span> <span class="nt">-d</span><span class="s1">'/'</span><span class="k">)</span><span class="p">;</span><span class="k">do
</span>sjaracne <span class="k">${</span><span class="nv">i</span><span class="k">}</span> <span class="nv">$indir</span>/<span class="k">${</span><span class="nv">i</span><span class="k">}</span>/<span class="k">*</span>.exp <span class="nv">$indir</span>/<span class="k">${</span><span class="nv">i</span><span class="k">}</span>/tf/<span class="k">*</span>.txt <span class="nv">$indir</span>/<span class="k">${</span><span class="nv">i</span><span class="k">}</span>/tf/ <span class="nt">--c_threshold</span> 0.01<span class="p">;</span>
<span class="nb">echo</span> <span class="k">${</span><span class="nv">i</span><span class="k">}</span> has been submitted!<span class="p">;</span>
<span class="k">done</span>

</code></pre></div></div>

<h2 id="find-cell-type-specific-master-regulator">Find cell type specific master regulator</h2>

<p>Identify master regulator from content based network is the key step in scMINER to help understanding your scRNA-seq data.</p>

<h3 id="calculate-inferred-activity">Calculate Inferred activity</h3>
<p>TF acitivities are calculated by integrating expression profile of their targets. Targets identified from SJARACNe of perticular TF was normalized and averaged to infer TF activity.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">acs.12k</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GetActivityFromSJARANCE</span><span class="p">(</span><span class="w">
			 </span><span class="n">SJaracne_output_path</span><span class="o">=</span><span class="s2">"Sjaracne/"</span><span class="p">,</span><span class="w">
			 </span><span class="n">SJaracne_input_eset</span><span class="o">=</span><span class="n">eset.demo</span><span class="p">,</span><span class="w">
			 </span><span class="n">activity.method</span><span class="o">=</span><span class="s2">"unweighted"</span><span class="p">,</span><span class="w"> 
			 </span><span class="n">activity.norm</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> 
			 </span><span class="n">group_tag</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"celltype"</span><span class="p">,</span><span class="w">
			 </span><span class="n">save_network_file</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">save_path</span><span class="o">=</span><span class="kc">NA</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="find-differential-activity-tf">Find Differential activity TF</h3>

<p>The function <code class="highlighter-rouge">FindDAG</code> was designed for identify highly differentiated TF from SJARACNe inferred activity matrix. In order to do so, we did two sided student’s t-test to compare mean acitivty from one cell type V.S. the others.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DAG_result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">FindDAG</span><span class="p">(</span><span class="n">eset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">acs.demo</span><span class="p">,</span><span class="n">group_tag</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"celltype"</span><span class="p">)</span><span class="w">

</span></code></pre></div></div>
<p>This function will output a full matrix that contians all TF occurred in original dataset, statistics such as t.statistics, p-value, 95%CI, etc. are outputed to help idenify master regulators. You can save it in txt or xlsx for checking.</p>

<p>While you can also visualize top master regulator candidates in heatmap or violinplots. Only png as plotting device is supported.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">TF_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">TopMasterRegulator</span><span class="p">(</span><span class="n">DAG_result</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">DAG_result</span><span class="p">,</span><span class="w">
                              </span><span class="n">celltype</span><span class="o">=</span><span class="n">levels</span><span class="p">(</span><span class="n">acs.12k</span><span class="o">$</span><span class="n">celltype</span><span class="p">),</span><span class="w"> </span><span class="c1"># ensure cluster order</span><span class="w">
                              </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">degree_filter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">50</span><span class="p">,</span><span class="m">500</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gene_heatmap</span><span class="p">(</span><span class="n">eset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">acs.12k</span><span class="p">,</span><span class="n">target</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TF_list</span><span class="p">,</span><span class="n">group_tag</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"celltype"</span><span class="p">,</span><span class="w">
             </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="n">save_plot</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">cluster_rows</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w">
             </span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Activity"</span><span class="p">,</span><span class="n">plot_name</span><span class="o">=</span><span class="s2">"./21_TopTFHeatmap.png"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="./plots/4_1_TopTFHeatmap.png" alt="" /></p>

<p>You can also check some known master regulators as postivie control of your network analysis:</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">feature_vlnplot</span><span class="p">(</span><span class="n">eset</span><span class="o">=</span><span class="n">acs.12k</span><span class="p">,</span><span class="w"> 
		</span><span class="n">target</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"LEF1.TF"</span><span class="p">,</span><span class="s2">"TCF7.TF"</span><span class="p">,</span><span class="s2">"BATF.TF"</span><span class="p">,</span><span class="s2">"TCF7.TF"</span><span class="p">,</span><span class="s2">"TBX21.TF"</span><span class="p">,</span><span class="s2">"IRF8.TF"</span><span class="p">,</span><span class="s2">"SPIB.TF"</span><span class="p">,</span><span class="s2">"BATF3.TF"</span><span class="p">,</span><span class="s2">"CEBPA.TF"</span><span class="p">),</span><span class="w">
       		</span><span class="n">ylabel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Activity"</span><span class="p">,</span><span class="n">group_tag</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"celltype"</span><span class="p">,</span><span class="n">drawquantiles</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="./plots/4_2_Known_MR_vlnplot.png" alt="" /></p>

<p>In order to conduct network analysis utilizing SJARACNe generated cell type specific networks, please infer 	<code class="highlighter-rouge">Cell-type spefic network analysis tab</code>.</p>

<hr />
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## R session info


R version 3.5.0 (2018-04-23)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] bindrcpp_0.2.2      Biobase_2.42.0      BiocGenerics_0.28.0 dplyr_0.7.8        
[5] RColorBrewer_1.1-2  ggplot2_3.1.0       reshape2_1.4.3     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.0       rstudioapi_0.8   bindr_0.1.1      magrittr_1.5     tidyselect_0.2.5
 [6] munsell_0.5.0    colorspace_1.3-2 R6_2.3.0         rlang_0.3.0.1    stringr_1.3.1   
[11] plyr_1.8.4       tools_3.5.0      grid_3.5.0       gtable_0.2.0     withr_2.1.2     
[16] digest_0.6.18    yaml_2.2.0       lazyeval_0.2.1   assertthat_0.2.0 tibble_1.4.2    
[21] crayon_1.3.4     zip_1.0.0        purrr_0.2.5      glue_1.3.0       labeling_0.3    
[26] openxlsx_4.1.0   stringi_1.2.4    compiler_3.5.0   pillar_1.3.1     scales_1.0.0    
[31] pkgconfig_2.0.2 
</code></pre></div></div>
<hr />



          
        </div>
      </div>
    </div>
  </div>

</body>
</html>
