<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Actvity-based analysis | scMINER: a mutual information-based framework for identifying hidden drivers from single-cell omics data</title>
  <meta name="description" content="This is the full documentation of scMINER R package (v-1.1.0)." />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Actvity-based analysis | scMINER: a mutual information-based framework for identifying hidden drivers from single-cell omics data" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="This is the full documentation of scMINER R package (v-1.1.0)." />
  <meta name="github-repo" content="jyyulab/scMINER_documentation" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Actvity-based analysis | scMINER: a mutual information-based framework for identifying hidden drivers from single-cell omics data" />
  
  <meta name="twitter:description" content="This is the full documentation of scMINER R package (v-1.1.0)." />
  

<meta name="author" content="John Doe" />


<meta name="date" content="2024-08-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="network-inference.html"/>
<link rel="next" href="session-information.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><img src="images/scMINER_logo.png" alt="" width="250px"></a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#a-few-concepts"><i class="fa fa-check"></i><b>1.1</b> A few concepts</a>
<ul>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#sparseeset"><i class="fa fa-check"></i>SparseEset</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#mutual-information"><i class="fa fa-check"></i>Mutual Information</a></li>
<li class="chapter" data-level="" data-path="introduction.html"><a href="introduction.html#gene-activity"><i class="fa fa-check"></i>Gene Activity</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#why-use-scminer"><i class="fa fa-check"></i><b>1.2</b> Why use scMINER</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#citation"><i class="fa fa-check"></i><b>1.3</b> Citation</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#support"><i class="fa fa-check"></i><b>1.4</b> Support</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="get-started.html"><a href="get-started.html"><i class="fa fa-check"></i><b>2</b> Get started</a>
<ul>
<li class="chapter" data-level="2.1" data-path="get-started.html"><a href="get-started.html#installation"><i class="fa fa-check"></i><b>2.1</b> Installation</a>
<ul>
<li class="chapter" data-level="" data-path="get-started.html"><a href="get-started.html#install-scminer-r-package"><i class="fa fa-check"></i>Install scMINER R package</a></li>
<li class="chapter" data-level="" data-path="get-started.html"><a href="get-started.html#install-mica-and-sjaracne"><i class="fa fa-check"></i>Install MICA and SJARACNe</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="get-started.html"><a href="get-started.html#demo-data"><i class="fa fa-check"></i><b>2.2</b> Demo data</a></li>
<li class="chapter" data-level="2.3" data-path="get-started.html"><a href="get-started.html#create-project-space"><i class="fa fa-check"></i><b>2.3</b> Create project space</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="generate-gene-expresion-matrix.html"><a href="generate-gene-expresion-matrix.html"><i class="fa fa-check"></i><b>3</b> Generate gene expresion matrix</a>
<ul>
<li class="chapter" data-level="3.1" data-path="generate-gene-expresion-matrix.html"><a href="generate-gene-expresion-matrix.html#from-data-directory-by-10x-genomics"><i class="fa fa-check"></i><b>3.1</b> From data directory by 10x Genomics</a></li>
<li class="chapter" data-level="3.2" data-path="generate-gene-expresion-matrix.html"><a href="generate-gene-expresion-matrix.html#from-text-table-file"><i class="fa fa-check"></i><b>3.2</b> From text-table file</a></li>
<li class="chapter" data-level="3.3" data-path="generate-gene-expresion-matrix.html"><a href="generate-gene-expresion-matrix.html#from-hdf5-file-by-10x-genomics"><i class="fa fa-check"></i><b>3.3</b> From HDF5 file by 10x Genomics</a></li>
<li class="chapter" data-level="3.4" data-path="generate-gene-expresion-matrix.html"><a href="generate-gene-expresion-matrix.html#from-h5ad-file"><i class="fa fa-check"></i><b>3.4</b> From H5AD file</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="create-sparseeset-object.html"><a href="create-sparseeset-object.html"><i class="fa fa-check"></i><b>4</b> Create SparseEset object</a>
<ul>
<li class="chapter" data-level="4.1" data-path="create-sparseeset-object.html"><a href="create-sparseeset-object.html#solely-from-the-gene-expression-matrix"><i class="fa fa-check"></i><b>4.1</b> Solely from the gene expression matrix</a></li>
<li class="chapter" data-level="4.2" data-path="create-sparseeset-object.html"><a href="create-sparseeset-object.html#using-self-customized-meta-data"><i class="fa fa-check"></i><b>4.2</b> Using self-customized meta data</a></li>
<li class="chapter" data-level="4.3" data-path="create-sparseeset-object.html"><a href="create-sparseeset-object.html#from-multiple-samples"><i class="fa fa-check"></i><b>4.3</b> From multiple samples</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="data-filtration.html"><a href="data-filtration.html"><i class="fa fa-check"></i><b>5</b> Data filtration</a>
<ul>
<li class="chapter" data-level="5.1" data-path="data-filtration.html"><a href="data-filtration.html#qc-report"><i class="fa fa-check"></i><b>5.1</b> QC report</a></li>
<li class="chapter" data-level="5.2" data-path="data-filtration.html"><a href="data-filtration.html#filter-the-sparse-eset-object"><i class="fa fa-check"></i><b>5.2</b> Filter the sparse eset object</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="data-filtration.html"><a href="data-filtration.html#data-filtration-with-auto-mode"><i class="fa fa-check"></i><b>5.2.1</b> Data filtration with auto mode</a></li>
<li class="chapter" data-level="5.2.2" data-path="data-filtration.html"><a href="data-filtration.html#data-filtration-with-manual-mode"><i class="fa fa-check"></i><b>5.2.2</b> Data filtration with manual mode</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-normalization.html"><a href="data-normalization.html"><i class="fa fa-check"></i><b>6</b> Data normalization</a></li>
<li class="chapter" data-level="7" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html"><i class="fa fa-check"></i><b>7</b> MI-based clustering analysis</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#introduction-to-mica"><i class="fa fa-check"></i><b>7.1</b> Introduction to MICA</a></li>
<li class="chapter" data-level="7.2" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#generate-mica-input"><i class="fa fa-check"></i><b>7.2</b> Generate MICA input</a></li>
<li class="chapter" data-level="7.3" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#run-mica"><i class="fa fa-check"></i><b>7.3</b> Run MICA</a></li>
<li class="chapter" data-level="7.4" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#integrate-mica-outputs-into-sparseeset-object"><i class="fa fa-check"></i><b>7.4</b> Integrate MICA outputs into SparseEset object</a></li>
<li class="chapter" data-level="7.5" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#visualize-the-mica-output"><i class="fa fa-check"></i><b>7.5</b> Visualize the MICA output</a>
<ul>
<li class="chapter" data-level="7.5.1" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#color-coded-by-cluster-labels"><i class="fa fa-check"></i><b>7.5.1</b> Color-coded by cluster labels</a></li>
<li class="chapter" data-level="7.5.2" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#color-coded-by-true-label-of-cell-types"><i class="fa fa-check"></i><b>7.5.2</b> Color-coded by true label of cell types</a></li>
<li class="chapter" data-level="7.5.3" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#color-coded-by-numi-for-qc-purpose"><i class="fa fa-check"></i><b>7.5.3</b> Color-coded by nUMI, for QC purpose</a></li>
<li class="chapter" data-level="7.5.4" data-path="mi-based-clustering-analysis.html"><a href="mi-based-clustering-analysis.html#color-coded-by-nfeature-for-qc-purpose"><i class="fa fa-check"></i><b>7.5.4</b> Color-coded by nFeature, for QC purpose</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html"><i class="fa fa-check"></i><b>8</b> Cell type annotation</a>
<ul>
<li class="chapter" data-level="8.1" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#supervised-cell-type-annotation"><i class="fa fa-check"></i><b>8.1</b> Supervised cell type annotation</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#annotate-using-signature-scores"><i class="fa fa-check"></i><b>8.1.1</b> Annotate using signature scores</a></li>
<li class="chapter" data-level="8.1.2" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#annotate-using-individual-marker-genes"><i class="fa fa-check"></i><b>8.1.2</b> Annotate using individual marker genes</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#unsupervised-cell-type-annotation"><i class="fa fa-check"></i><b>8.2</b> Unsupervised cell type annotation</a></li>
<li class="chapter" data-level="8.3" data-path="cell-type-annotation.html"><a href="cell-type-annotation.html#add-cell-type-annotations-to-sparseexpressionset-object"><i class="fa fa-check"></i><b>8.3</b> Add cell type annotations to SparseExpressionSet object</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="network-inference.html"><a href="network-inference.html"><i class="fa fa-check"></i><b>9</b> Network inference</a>
<ul>
<li class="chapter" data-level="9.1" data-path="network-inference.html"><a href="network-inference.html#generate-sjaracne-input-files"><i class="fa fa-check"></i><b>9.1</b> Generate SJARACNe input files</a></li>
<li class="chapter" data-level="9.2" data-path="network-inference.html"><a href="network-inference.html#run-sjaracne"><i class="fa fa-check"></i><b>9.2</b> Run SJARACNe</a></li>
<li class="chapter" data-level="9.3" data-path="network-inference.html"><a href="network-inference.html#assess-the-quality-of-networks"><i class="fa fa-check"></i><b>9.3</b> Assess the quality of networks</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="network-inference.html"><a href="network-inference.html#introduction-to-the-network-file-by-sjaracne"><i class="fa fa-check"></i><b>9.3.1</b> Introduction to the network file by SJARACNe</a></li>
<li class="chapter" data-level="9.3.2" data-path="network-inference.html"><a href="network-inference.html#generate-network-qc-report"><i class="fa fa-check"></i><b>9.3.2</b> Generate network QC report</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="actvity-based-analysis.html"><a href="actvity-based-analysis.html"><i class="fa fa-check"></i><b>10</b> Actvity-based analysis</a>
<ul>
<li class="chapter" data-level="10.1" data-path="actvity-based-analysis.html"><a href="actvity-based-analysis.html#calculate-the-activities"><i class="fa fa-check"></i><b>10.1</b> Calculate the activities</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="actvity-based-analysis.html"><a href="actvity-based-analysis.html#calculate-activities-per-group"><i class="fa fa-check"></i><b>10.1.1</b> Calculate activities per group</a></li>
<li class="chapter" data-level="10.1.2" data-path="actvity-based-analysis.html"><a href="actvity-based-analysis.html#calculate-activities-in-batch"><i class="fa fa-check"></i><b>10.1.2</b> Calculate activities in batch</a></li>
<li class="chapter" data-level="10.1.3" data-path="actvity-based-analysis.html"><a href="actvity-based-analysis.html#save-activity-eset-object"><i class="fa fa-check"></i><b>10.1.3</b> Save activity eset object</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="actvity-based-analysis.html"><a href="actvity-based-analysis.html#differential-activity-analysis"><i class="fa fa-check"></i><b>10.2</b> Differential activity analysis</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="session-information.html"><a href="session-information.html"><i class="fa fa-check"></i><b>11</b> Session information</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">scMINER: a mutual information-based framework for identifying hidden drivers from single-cell omics data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="actvity-based-analysis" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">Chapter 10</span> Actvity-based analysis<a href="actvity-based-analysis.html#actvity-based-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The <strong>driver activity estimation</strong> is one of the most important features of scMINER. <strong>Mathematically</strong>, the activity of one driver is a type of mean of the expressions of its targets. And <strong>biologically</strong>, the activity can be interpreted as a measure that describes how actively the driver functions, like the enzymes in digesting their subtracts, kinase in activating their downstream genes. Given the gene expression profiles and networks, scMINER can estimate the activities of some predefined drivers, including not only transcription factors (TFs) but also signaling genes (SIGs). scMINER provides a few functions to effortlessly calculate the activities, identify the hidden drivers and visualize them in multiple ways.</p>
<div id="calculate-the-activities" class="section level2 hasAnchor" number="10.1">
<h2><span class="header-section-number">10.1</span> Calculate the activities<a href="actvity-based-analysis.html#calculate-the-activities" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>scMINER provides two functions, <code>getActivity_individual()</code> and <code>getActivity_inBatch()</code>, to calculate the driver activities.</p>
<div id="calculate-activities-per-group" class="section level3 hasAnchor" number="10.1.1">
<h3><span class="header-section-number">10.1.1</span> Calculate activities per group<a href="actvity-based-analysis.html#calculate-activities-per-group" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>getActivity_individual()</code> is designed to calculate the activities per group. It takes the network files as the input:</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="actvity-based-analysis.html#cb124-1" tabindex="-1"></a><span class="do">## let&#39;s use B cell as an example</span></span>
<span id="cb124-2"><a href="actvity-based-analysis.html#cb124-2" tabindex="-1"></a>activity_B.eset <span class="ot">&lt;-</span> <span class="fu">getActivity_individual</span>(<span class="at">input_eset =</span> pbmc14k_log2cpm.eset[, <span class="fu">pData</span>(pbmc14k_log2cpm.eset)<span class="sc">$</span>trueLabel <span class="sc">==</span> <span class="st">&quot;B&quot;</span>],</span>
<span id="cb124-3"><a href="actvity-based-analysis.html#cb124-3" tabindex="-1"></a>                                          <span class="at">network_file.tf =</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/demo_pbmc14k/PBMC14k/SJARACNe/B/TF/bt100_pc001/consensus_network_ncol_.txt&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scMINER&quot;</span>),</span>
<span id="cb124-4"><a href="actvity-based-analysis.html#cb124-4" tabindex="-1"></a>                                          <span class="at">network_file.sig =</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/demo_pbmc14k/PBMC14k/SJARACNe/B/SIG/bt100_pc001/consensus_network_ncol_.txt&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scMINER&quot;</span>),</span>
<span id="cb124-5"><a href="actvity-based-analysis.html#cb124-5" tabindex="-1"></a>                                          <span class="at">driver_type =</span> <span class="st">&quot;TF_SIG&quot;</span>)</span></code></pre></div>
</div>
<div id="calculate-activities-in-batch" class="section level3 hasAnchor" number="10.1.2">
<h3><span class="header-section-number">10.1.2</span> Calculate activities in batch<a href="actvity-based-analysis.html#calculate-activities-in-batch" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>If you need to calculate the activity for multiple groups, this is usually the case, you can do it using <code>getActivity_individual()</code> as shown above one by one and merge the esets after that. Or, scMINER privides another function, <code>getActivity_inBatch()</code>, to calculate the activity in batch:</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="actvity-based-analysis.html#cb125-1" tabindex="-1"></a><span class="do">## let&#39;s use B cell as an example</span></span>
<span id="cb125-2"><a href="actvity-based-analysis.html#cb125-2" tabindex="-1"></a>activity.eset <span class="ot">&lt;-</span> <span class="fu">getActivity_inBatch</span>(<span class="at">input_eset =</span> pbmc14k_log2cpm.eset, <span class="at">sjaracne_dir =</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata/demo_pbmc14k/PBMC14k/SJARACNe&quot;</span>, <span class="at">package =</span> <span class="st">&quot;scMINER&quot;</span>),  <span class="at">group_name =</span> <span class="st">&quot;trueLabel&quot;</span>, <span class="at">driver_type =</span> <span class="st">&quot;TF_SIG&quot;</span>, <span class="at">activity_method =</span> <span class="st">&quot;mean&quot;</span>, <span class="at">do.z_normalization =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## 7 groups were found in trueLabel ...
## Checking network files for each group ...
##  Group 1 / 7 : Monocyte ...
##      TF network check passed!
##      SIG network check passed!
##  Group 2 / 7 : B ...
##      TF network check passed!
##      SIG network check passed!
##  Group 3 / 7 : CD4Treg ...
##      TF network check passed!
##      SIG network check passed!
##  Group 4 / 7 : CD4TN ...
##      TF network check passed!
##      SIG network check passed!
##  Group 5 / 7 : CD4TCM ...
##      TF network check passed!
##      SIG network check passed!
##  Group 6 / 7 : NK ...
##      TF network check passed!
##      SIG network check passed!
##  Group 7 / 7 : CD8TN ...
##      TF network check passed!
##      SIG network check passed!
## Calculating activity for each group ...
##  Group 1 / 7 : Monocyte ...
##  Activity calculation is completed successfully!
##  Group 2 / 7 : B ...
##  Activity calculation is completed successfully!
##  Group 3 / 7 : CD4Treg ...
##  Activity calculation is completed successfully!
##  Group 4 / 7 : CD4TN ...
##  Activity calculation is completed successfully!
##  Group 5 / 7 : CD4TCM ...
##  Activity calculation is completed successfully!
##  Group 6 / 7 : NK ...
##  Activity calculation is completed successfully!
##  Group 7 / 7 : CD8TN ...
##  Activity calculation is completed successfully!
## NAs were found in the activity matrix and have been replaced by the minimum value:  -0.3968794 .</code></pre>
</div>
<div id="save-activity-eset-object" class="section level3 hasAnchor" number="10.1.3">
<h3><span class="header-section-number">10.1.3</span> Save activity eset object<a href="actvity-based-analysis.html#save-activity-eset-object" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="actvity-based-analysis.html#cb127-1" tabindex="-1"></a><span class="fu">saveRDS</span>(activity.eset, <span class="at">file =</span> <span class="st">&quot;/your-path/PBMC14k/DATA/activity.eset&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="differential-activity-analysis" class="section level2 hasAnchor" number="10.2">
<h2><span class="header-section-number">10.2</span> Differential activity analysis<a href="actvity-based-analysis.html#differential-activity-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Similar to <code>getDE()</code>, scMINER provides a function, <code>getDA()</code>, to perform the differential activity analysis and identify the group-specific drivers.</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="actvity-based-analysis.html#cb128-1" tabindex="-1"></a><span class="do">## 1. To perform differential expression analysis in a 1-vs-rest manner for all groups</span></span>
<span id="cb128-2"><a href="actvity-based-analysis.html#cb128-2" tabindex="-1"></a>da_res1 <span class="ot">&lt;-</span> <span class="fu">getDA</span>(<span class="at">input_eset =</span> activity.eset, <span class="at">group_by =</span> <span class="st">&quot;cell_type&quot;</span>, <span class="at">use_method =</span> <span class="st">&quot;t.test&quot;</span>)</span></code></pre></div>
<pre><code>## 7 groups were found in group_by column [ cell_type ].
## Since no group was specified, the differential analysis will be conducted among all groups in the group_by column [ cell_type ] in the 1-vs-rest manner.
##   1 / 7 : group 1 ( B ) vs the rest...
##   1912 cells were found for g1.
##   11693 cells were found for g0.
##   2 / 7 : group 1 ( CD4TCM ) vs the rest...
##   2022 cells were found for g1.
##   11583 cells were found for g0.
##   3 / 7 : group 1 ( CD4TN ) vs the rest...
##   2505 cells were found for g1.
##   11100 cells were found for g0.
##   4 / 7 : group 1 ( CD4Treg ) vs the rest...
##   1448 cells were found for g1.
##   12157 cells were found for g0.
##   5 / 7 : group 1 ( CD8TN ) vs the rest...
##   2014 cells were found for g1.
##   11591 cells were found for g0.
##   6 / 7 : group 1 ( Monocyte ) vs the rest...
##   1786 cells were found for g1.
##   11819 cells were found for g0.
##   7 / 7 : group 1 ( NK ) vs the rest...
##   1918 cells were found for g1.
##   11687 cells were found for g0.</code></pre>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="actvity-based-analysis.html#cb130-1" tabindex="-1"></a><span class="fu">head</span>(da_res1)</span></code></pre></div>
<pre><code>##       feature g1_tag                                 g0_tag       g1_avg
## 4   AASDH_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.008071658
## 6    AATF_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.051767485
## 12  ABCB8_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.077615607
## 8   ABCA2_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.081643603
## 10  ABCB1_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.134357577
## 3  AARSD1_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.126010447
##        g0_avg     g1_pct     g0_pct       log2FC          Pval           FDR
## 4  -0.1025141 0.43043933 0.13991277  0.094442475 2.225074e-308  0.000000e+00
## 6  -0.1084165 0.21652720 0.08757376  0.056649005 3.918924e-189 5.878386e-189
## 12 -0.1094585 0.35251046 0.14153767  0.031842866  3.623209e-12  3.952592e-12
## 8  -0.1101867 0.10198745 0.15676045  0.028543082  1.914570e-58  2.418404e-58
## 10 -0.1559384 0.04393305 0.06114770  0.021580866  8.079661e-27  9.233898e-27
## 3  -0.1225746 0.04184100 0.08192936 -0.003435892  4.213744e-02  4.213744e-02
##      Zscore
## 4  37.53784
## 6  29.33316
## 12  6.95115
## 8  16.11775
## 10 10.72137
## 3  -2.03216</code></pre>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="actvity-based-analysis.html#cb132-1" tabindex="-1"></a><span class="do">## 2. To perform differential expression analysis in a 1-vs-rest manner for one specific group</span></span>
<span id="cb132-2"><a href="actvity-based-analysis.html#cb132-2" tabindex="-1"></a>da_res2 <span class="ot">&lt;-</span> <span class="fu">getDA</span>(<span class="at">input_eset =</span> activity.eset, <span class="at">group_by =</span> <span class="st">&quot;cell_type&quot;</span>, <span class="at">g1 =</span> <span class="fu">c</span>(<span class="st">&quot;B&quot;</span>), <span class="at">use_method =</span> <span class="st">&quot;t.test&quot;</span>)</span>
<span id="cb132-3"><a href="actvity-based-analysis.html#cb132-3" tabindex="-1"></a></span>
<span id="cb132-4"><a href="actvity-based-analysis.html#cb132-4" tabindex="-1"></a><span class="do">## 3. To perform differential expression analysis in a rest-vs-1 manner for one specific group</span></span>
<span id="cb132-5"><a href="actvity-based-analysis.html#cb132-5" tabindex="-1"></a>da_res3 <span class="ot">&lt;-</span> <span class="fu">getDA</span>(<span class="at">input_eset =</span> activity.eset, <span class="at">group_by =</span> <span class="st">&quot;cell_type&quot;</span>, <span class="at">g0 =</span> <span class="fu">c</span>(<span class="st">&quot;B&quot;</span>), <span class="at">use_method =</span> <span class="st">&quot;t.test&quot;</span>)</span>
<span id="cb132-6"><a href="actvity-based-analysis.html#cb132-6" tabindex="-1"></a></span>
<span id="cb132-7"><a href="actvity-based-analysis.html#cb132-7" tabindex="-1"></a><span class="do">## 4. To perform differential expression analysis in a 1-vs-1 manner for any two groups</span></span>
<span id="cb132-8"><a href="actvity-based-analysis.html#cb132-8" tabindex="-1"></a>da_res4 <span class="ot">&lt;-</span> <span class="fu">getDA</span>(<span class="at">input_eset =</span> activity.eset, <span class="at">group_by =</span> <span class="st">&quot;cell_type&quot;</span>, <span class="at">g1 =</span> <span class="fu">c</span>(<span class="st">&quot;CD4Treg&quot;</span>), <span class="at">g0 =</span> <span class="fu">c</span>(<span class="st">&quot;CD4TCM&quot;</span>), <span class="at">use_method =</span> <span class="st">&quot;t.test&quot;</span>)</span></code></pre></div>
<p>The <code>getTopFeatures()</code> function can aslo be used to easily extract the group-specific markers from the differential expression result:</p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="actvity-based-analysis.html#cb133-1" tabindex="-1"></a>top_drivers <span class="ot">&lt;-</span> <span class="fu">getTopFeatures</span>(<span class="at">input_table =</span> da_res1, <span class="at">number =</span> <span class="dv">10</span>, <span class="at">group_by =</span> <span class="st">&quot;g1_tag&quot;</span>, <span class="at">sort_by =</span> <span class="st">&quot;log2FC&quot;</span>, <span class="at">sort_decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb133-2"><a href="actvity-based-analysis.html#cb133-2" tabindex="-1"></a><span class="fu">dim</span>(top_drivers)</span></code></pre></div>
<pre><code>## [1] 16 11</code></pre>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="actvity-based-analysis.html#cb135-1" tabindex="-1"></a><span class="fu">head</span>(top_drivers)</span></code></pre></div>
<pre><code>##       feature g1_tag                                 g0_tag       g1_avg
## 4   AASDH_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.008071658
## 6    AATF_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.051767485
## 12  ABCB8_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.077615607
## 8   ABCA2_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.081643603
## 10  ABCB1_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.134357577
## 3  AARSD1_SIG      B CD4TCM,CD4TN,CD4Treg,CD8TN,Monocyte,NK -0.126010447
##        g0_avg     g1_pct     g0_pct       log2FC          Pval           FDR
## 4  -0.1025141 0.43043933 0.13991277  0.094442475 2.225074e-308  0.000000e+00
## 6  -0.1084165 0.21652720 0.08757376  0.056649005 3.918924e-189 5.878386e-189
## 12 -0.1094585 0.35251046 0.14153767  0.031842866  3.623209e-12  3.952592e-12
## 8  -0.1101867 0.10198745 0.15676045  0.028543082  1.914570e-58  2.418404e-58
## 10 -0.1559384 0.04393305 0.06114770  0.021580866  8.079661e-27  9.233898e-27
## 3  -0.1225746 0.04184100 0.08192936 -0.003435892  4.213744e-02  4.213744e-02
##      Zscore
## 4  37.53784
## 6  29.33316
## 12  6.95115
## 8  16.11775
## 10 10.72137
## 3  -2.03216</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="network-inference.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="session-information.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/10-activity-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
