---
layout: default
title: Home
nav_order: 1
permalink: /
---

<img src="docs/plots/scMINER_V4.png" alt="Logo" width="200" height="300"/>

**S**ingle-**C**ell **M**utual **I**nformation-based **N**etwork **E**ngineering **R**anger
{: .fs-6 .fw-300 }

[Get started now](#getting-started){: .btn .btn-primary .fs-5 .mb-4 .mb-md-0 .mr-2 } 
[View it on GitHub](https://github.com/jyyulab/scMINER){: .btn .fs-5 }

---

{:.no_toc}

## Overview
**scMINER** is a **systems biology** toolkit for QC and analysis of high-throughput single cell RNA-seq data. 
For the better understanding of the sources of heterogeneity from single-cell RNA-seq data, scMINER enable users 
to identify and interpret **cell type-specific hidden drivers**, where the drivers can be either transcription 
factors (TF) or signaling factors (SIG). 

**scMINER** also includes new mutual information based approach [MICA](https://github.com/jyyulab/MICA) for classifying 
cells and improved scalable version of [SJARACNe](https://github.com/jyyulab/SJARACNe) for cell type-specific 
gene network reverse engineering.

<img src="docs/plots/scMINER_Overview.png" alt="Overview"/> 

## Citing scMINER
If scMINER is useful to your work, please cite the following paper:

scMINER bioRxiv version first?


## Contact
The scMINER software is developed by [Yu Lab](https://www.stjude.org/directory/y/jiyang-yu.html) at [St. Jude 
Children's Research Hospital](https://www.stjude.org/). 

Weâ€™re very interested to hear feedback about using scMINER in your analysis. If you find any issues/bugs, or 
have any suggestions, please don't hesitate to create issues in [GitHub](https://github.com/jyyulab/scMINER/issues), 
or contact us directly. 
