<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Liang (Adam) Ding, Hao Shi">
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>Basic analysis on PBMC (14k) - scMINER</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../../css/theme.css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Basic analysis on PBMC (14k)";
    var mkdocs_page_input_path = "tutorials/PBMC-14k.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/languages/yaml.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  <script async src="https://www.googletagmanager.com/gtag/js?id=G-ABC123"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'G-ABC123');
  </script>
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> scMINER</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../install/">Installation</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="" href="../../api.md">API</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../reference/">References</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../../contact/">Contact</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Tutorials</span></p>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">Basic analysis on PBMC (14k)</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#data-loading-and-preprocessing">Data loading and preprocessing</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#read-10x-genomics-data">Read 10x genomics data</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#quality-control-and-data-filtering">Quality control and data filtering</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#normalization-and-transformation">Normalization and transformation</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#perform-clustering-analysis-via-mica">Perform clustering analysis via MICA</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#generate-mica-input">Generate MICA input</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#run-mica-from-the-command-line">Run MICA from the command line</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#read-mica-output">Read MICA output</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#cell-type-annotation-after-clustering">Cell type annotation after clustering</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#marker-gene-visualization">Marker gene visualization</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#cell-type-annotation">Cell type annotation</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#network-generation-via-sjaracne">Network generation via SJARACNe</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#generate-sjaracne-input">Generate SJARACNe input</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#run-sjaracne-from-the-command-line">Run SJARACNe from the command line</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#identify-cell-type-specific-hidden-drivers-via-minie">Identify cell-type-specific hidden drivers via MINIE</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#calculate-activity">Calculate activity</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#driver-estimation-by-differential-activity-analysis">Driver estimation by differential activity analysis</a>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../CD8T/">KO vs. WT analysis on CD8+ T cells</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../MICA-advanced/">Mutual information-based clustering analysis (MICA)</a>
                    </li>
                    <li class="toctree-l1"><a class="" href="../network-visual.md">Network visualization</a>
                    </li>
                    <li class="toctree-l1"><a class="" href="../function-analysis.md">Driver target function analysis</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../interoperability/">Interoperability between SparseEset, Seurat, and anndata</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">scMINER</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
        
          <li>Tutorials &raquo;</li>
        
      
    
    <li>Basic analysis on PBMC (14k)</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/jyyulab/scMINER/edit/master/docs/tutorials/PBMC-14k.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>

          <div role="main">
            <div class="section">
              
                <h1 id="scminer-guided-analysis-on-14k-pbmcs-from-10x-genomics">scMINER Guided Analysis on 14k PBMCs from 10x Genomics</h1>
<p>This tutorial introduce you scMINER's basic analysis using a PBMC dataset with 10 sorted populations of 2k cells per population [<a href="../../reference/#[Zheng et al., 2017]">Zheng et al., 2017</a>]. We remove CD34+ cells, CD4+ helper T cells, and total CD8+ cytotoxic cells from the dataset because of low sorting purity or a significant overlap with other immune cells based on the sorting strategy, which creates a new dataset with seven known cell types and 14k cells in total. The original dataset is freely available under accession number <a href="https://www.ncbi.nlm.nih.gov/sra?term=SRP073767">SRP073767</a> and <a href="https://zenodo.org/record/3357167#.YhQNF2RKj6V">Zenodo</a>.</p>
<h2 id="data-loading-and-preprocessing">Data loading and preprocessing</h2>
<h3 id="read-10x-genomics-data">Read 10x genomics data</h3>
<p>We start by reading the downsampled 10x genomics data with function <code>readscRNAseqData()</code>. This function reads data from either 10x genomics standard output files (usually contains three individual files: matrix.mtx, barcodes.tsv, features.tsv) or other text files by passing arguments to <code>read.delim()</code>. The function creates a <code>SparseExpressionSet</code> object adapted from <code>ExpressionSet</code> class if set <code>CreateSparseEset=T</code>; otherwise, it creates a list object that stores expression data, feature data and sample data under three separate slots. If <code>add.meta=T</code>, then additional sample info such as total number of UMI will be calculated and stored in the object. The following command sets <code>is.10x=T</code>, <code>CreateSparseEset = T</code>, and <code>add.meta=T</code>.</p>
<p>To load data from standard 10x genomics files, creating a <code>SparseEset</code> directly is recommended.</p>
<pre><code class="language-R"># Create a SparseExpressionSet object with meta info
pbmc.14k.eset &lt;- readscRNAseqData(file = &quot;./PBMC14k_input/&quot;, is.10x = T, CreateSparseEset = T, 
                                  add.meta = T)
</code></pre>
<p>The following code loads data that are not from standard 10x genomics files.</p>
<pre><code class="language-R"># Read the data as a data.frame (not create SparseExpressionSet object or add meta info)
pbmc.14k &lt;- readscRNAseqData(file=&quot;./PBMC14k_input/PBMC_20k_MICA_input_filter_14k.txt&quot;, 
                             is.10x = F, CreateSparseEset = F, add.meta = F, sep='\t')
</code></pre>
<p>Then a SparseEset object can be created by using <code>CreateSparseEset()</code> function.</p>
<pre><code class="language-R"># Set add.meta=T to run the quality control and store the info in the object
pbmc.14k.eset &lt;- CreateSparseEset(data = t(as.matrix(pbmc.14k)), add.meta = T)

# (Optional) Define your group information. If there is no additional sample info available, 
# just put a character here
pbmc.14k.eset$group &lt;- sapply(strsplit(pbmc.14k.eset$cellName, &quot;-&quot;), &quot;[&quot;,2)
</code></pre>
<h3 id="quality-control-and-data-filtering">Quality control and data filtering</h3>
<p>Quality control assessments can be performed using <code>draw.scRNAseq.QC()</code> function, which outputs a html report generated through Rmarkdown <a href="../htmls/PBMC14k_scRNAseq_QC.html">PBMC14K_QC_report.html</a>. The report includes three essential quality control figures at both gene and cell levels. Suggested cutoffs will be returned as a list if set <code>output.cutoff=TRUE</code>.</p>
<pre><code class="language-R">cutoffs &lt;- draw.scRNAseq.QC(SparseEset = eset.14k, project.name = &quot;PBMC14k&quot;, plot.dir = &quot;./QC/&quot;,
                            group = &quot;group&quot;, # the meta info to group violin plots
                            output.cutoff = TRUE) # Output suggested cutoffs
</code></pre>
<p>The histogram plot visualizes a distribution of expressed genes among each cells. The blue vertical line shows the recommended cutoff (0.5% * number of cells) to filter lowly expressed cells.</p>
<p><img alt="QC_histogram_plot" src="../../images/1_1_Gene_QCmetrics_before_filtering.png" /></p>
<p>The violin plot visualizes the total UMI count and the total number of genes expressed in violin plots. Horizontal blue line indicates suggested high/low cutoffs. Suggested thresholds were computed based on Median Â± 3 * MAD (maximum absolute deviance). Suggested threshold numbers are also printed right above blue lines in labels.</p>
<p><img alt="QC_violin_plot" src="../../images/1_2_Cell_QC_1.png" /></p>
<p>The scatter plot shows mitochondrial gene expression percentage, and spike-in genes expression percentage (if spike-in genes exist) for each cell vs. the total number of UMIs. Cells with a high percentage of mitochondrial gene expression but a low total number of UMI counts are often considered as low quality cells (above the horizontal line).</p>
<p><img alt="QC_mito_plot" src="../../images/1_3_Cell_QC_2.png" /></p>
<p>We recommend using the thresholds in <code>cutoffs</code> list returned by <code>draw.scRNAseq.QC()</code> function. However, users can adjust the thresholds manually, e.g., <code>cutoffs$umi_cf_hi &lt;- Inf</code> will not filter on outliers caused by a high total UMI values.</p>
<pre><code class="language-R"># Perform the actual filtering
pbmc.14k.eset.filter &lt;- preMICA.filtering(SparseEset = pbmc.14k.eset, cutoffs = cutoffs) 
</code></pre>
<h3 id="normalization-and-transformation">Normalization and transformation</h3>
<p>We recommend count per million reads (CPM) normalization and log2 transformation (required) for performing MICA clustering analysis.</p>
<pre><code class="language-R">norm = 1e6 
exp.norm &lt;- sweep(exprs(pbmc.14k.eset.filter), 2, 
                  norm/unname(Matrix::colSums(exprs(pbmc.14k.eset.filter))), '*')

# log transformation (required by MICA)
exp.log2 &lt;- log(exp.norm + 1, base = 2)

# save as SparseEset
pbmc.14k.eset.log2 &lt;- CreateSparseEset(data = exp.log2, meta.data = pData(pbmc.14k.eset.filter), 
                                       feature.data = fData(pbmc.14k.eset.filter), add.meta = F)
</code></pre>
<h2 id="perform-clustering-analysis-via-mica">Perform clustering analysis via MICA</h2>
<p>After reviewing all the QC plots and finishing filtering, it is ready to perform clustering analysis. MICA is implemented in Python as a standalone command line tool. Thus we suggest saving your working directory and R session prior to running MICA.</p>
<h3 id="generate-mica-input">Generate MICA input</h3>
<p><code>generateMICAinput()</code> function takes a log transformed expression matrix as input, and outputs a cell by gene matrix stored in .h5 (default) or .txt file as input to MICA.</p>
<pre><code class="language-R"># Prepare MICA input
generateMICAinput(d = exp.log2 , filename = &quot;./PBMC14k_MICA_input.h5ad&quot;)

# (optional) Clean working environment
rm(exp.log2)
rm(exp.norm)
</code></pre>
<h3 id="run-mica-from-the-command-line">Run MICA from the command line</h3>
<p>Pick a MICA command to run in a command line environment based on the number of cells of your dataset. For dataset with less than <code>5k</code> cells, MICA MDS mode is recommended.</p>
<pre><code class="language-bash">mica mds -i ./PBMC14k_MICA_input.h5ad -o ./outputs -pn PBMC14k -nc 8 9 10 -dk 14 13 14 15 16 17 18 19
</code></pre>
<p>Here, <code>-pn</code> specifies a project name for naming the output files; <code>-nc</code> is an array of integers delimited by a single space, where each integer specifies a <code>k</code> to perform a k-mean clustering; <code>-dk</code> can be an integer or an array of integers delimited by a single space (default is <code>19</code>), it specifies the number of dimensions used in k-mean clusterings. Use <code>mica mds -h</code> to see more options of MICA MDS mode.</p>
<p>For datasets with more than <code>5k</code> cells, MICA GE mode is recommended.</p>
<pre><code class="language-bash">mica ge -i ./PBMC14k_MICA_input.h5ad -o ./outputs -ar 4.0 -nw 10 -nnm 100 -pdm 3.0
</code></pre>
<p>Here, <code>-ar</code> determines the maximum size of the communities (default: 3.0); <code>-nw</code> specifies the number of workers to run in parallel (default: 10); <code>-nnm</code> is the number of neighbors to build mutual information-based nearest neighbor graph (default 100); <code>-pdm</code> stands for pruning degree multiplier, the vertex with degree greater than pruning_degree_multi * num_neighbors_mi will be pruned. Use <code>mica ge -h</code> to find more options of MICA MDS mode.</p>
<p>A more detailed description of important parameters can be found in <a href="">API</a> page.</p>
<h3 id="read-mica-output">Read MICA output</h3>
<p>Function <code>readMICAoutput()</code> adds MICA clustering results to the <code>SparseEset</code> object created before, which will be the major data structure for downstream analysis. <strong>Note: All the functions for downstream analysis are compatible with both <code>expressionSet</code> and <code>SparseExpressionSet</code></strong>. </p>
<p>After choosing a clustering result based on known cell type signatures or silhouette scores (provided by GE mode). The clustering labels can be saved to <code>eset$ClusterRes</code> by setting <code>load_ClusterRes</code> as TRUE.</p>
<pre><code class="language-R"># Load clustering results
pbmc.14k.eset.log2 &lt;- readMICAoutput(eset = pbmc.14k.eset.log2, load_ClusterRes = TRUE, 
                                     output_file = &quot;MICA/clustering_UMAP_euclidean_24_1.82212.txt&quot;)
</code></pre>
<p>Note in <code>clustering_UMAP_euclidean_24_1.82212.txt</code>, <code>24</code> is the number of reduced dimensions, <code>1.82212</code> is the resolution used for Louvain clustering.</p>
<p><code>MICAplot()</code> function can be used to visualize MICA clustering results using <code>ggplot</code>.</p>
<pre><code class="language-R"># X, Y, color_by specify coordinate and clustering label entries in the eset phenoData; 
# pct is the size of the point
MICAplot(input_eset = pbmc.14k.eset.log2, X = &quot;X&quot;, Y = &quot;Y&quot;, color_by = &quot;ClusterRes&quot;, pct = 0.5)
</code></pre>
<p><img alt="PBMC_14k_UMAP" src="../../images/clustering_UMAP_euclidean_24_1.82212_ggplot.png" /></p>
<h2 id="cell-type-annotation-after-clustering">Cell type annotation after clustering</h2>
<h3 id="marker-gene-visualization">Marker gene visualization</h3>
<p>To help with identifing the optimal number of clusters and annotating the clusters, selected marker genes can be visualized with UMAP/t-SNE scatterplots, violin plots or heatmap plots via function <code>feature_highlighting()</code>, <code>feature_vlnplot()</code> and <code>feature_heatmap()</code> respectively.</p>
<pre><code class="language-R">genes_of_interest &lt;-c(&quot;CD3D&quot;, &quot;CD27&quot;, &quot;IL7R&quot;, &quot;SELL&quot;, &quot;CCR7&quot;, &quot;IL32&quot;, &quot;GZMA&quot;, &quot;GZMK&quot;, &quot;DUSP2&quot;, 
                      &quot;CD8A&quot;, &quot;GZMH&quot;, &quot;GZMB&quot;, &quot;CD79A&quot;, &quot;CD79B&quot;, &quot;CD86&quot;, &quot;CD14&quot;)

# UMAP scatter plot
feature_highlighting(input_eset = pbmc.14k.eset.log2, target = genes_of_interest, 
                     feature = &quot;geneSymbol&quot;, ylabel = &quot;log2Exp&quot;, x = &quot;X&quot;, y = &quot;Y&quot;, pct.size = 0.5)
</code></pre>
<p><img alt="PBMC_14k_UMAP" src="../../images/markers_UMAP.png" /></p>
<pre><code class="language-R"># Violoin plot
feature_vlnplot(input_eset = pbmc.14k.eset.log2, target = genes_of_interest, 
                feature = &quot;geneSymbol&quot;, group_by = &quot;ClusterRes&quot;, ylabel = &quot;log2Exp&quot;, ncol = 4)
</code></pre>
<p><img alt="PBMC_14k_violin" src="../../images/markers_violin.png" /></p>
<pre><code class="language-R"># Heatmap plot
feature_heatmap(input_eset = pbmc.14k.eset.log2, target = genes_of_interest, group_name = &quot;ClusterRes&quot;,
                save_plot = FALSE, width = 6, height = 6, name = &quot;log2Exp&quot;)
</code></pre>
<p><img alt="PBMC_14k_violin" src="../../images/markers_heatmap.png" /></p>
<h3 id="cell-type-annotation">Cell type annotation</h3>
<p>With a pre-defined marker gene list, <code>marker_bbplot()</code> function calculates a cell type signature score (weighted mean using the given weights in the marker list) for each cluster, and visualize all the signature scores using a bubble plot. Here, we curated a list of well-known marker genes of 9 common immune cell types and use these genes to calculate a marker score for each of the predicted PBMC clusters. <strong>Note: the pre-defined marker gene list has a required format. See <code>head(makers)</code> below</strong>.</p>
<pre><code class="language-R"># Read marker genes from an excel file
markers &lt;- read.xlsx(&quot;Immune_signatures.xlsx&quot;)

# Maker gene file has 3 required columns
head(markers)
  celltype markers weight
1 Monocyte    CD14      1
2 Monocyte  FCGR3A      1
3 Monocyte  FCGR3B      1
4 Monocyte   PTPRC      1
5 Monocyte  S100A8      1
6 Monocyte  S100A9      1

# Draw a bubble plot
draw.marker.bbp(ref = markers, input_eset = pbmc.14k.eset.log2, width = 6, height = 4, 
                feature = &quot;geneSymbol&quot;, group_name = &quot;ClusterRes&quot;, save_plot = FALSE)
</code></pre>
<p><img alt="PBMC_14k_violin" src="../../images/signature_score.png" /></p>
<p>Before diving into network generation section, assign the annotated cell types as factors in your expression set for keeping the current order of the clusters in following visualizations. <strong>Please do not include "_" in your cell type names as it would cause mis-parsing in the following analysis.</strong>.</p>
<pre><code class="language-R">indx &lt;- factor(x=c(&quot;Monocyte&quot;, &quot;CD8NaiveCTL&quot;, &quot;NK&quot;, &quot;Bcell&quot;, &quot;CD4TReg&quot;, &quot;CD4TCM&quot;, &quot;CD4NaiveT&quot;),
              levels=c(&quot;Monocyte&quot;, &quot;CD8NaiveCTL&quot;, &quot;NK&quot;, &quot;Bcell&quot;, &quot;CD4TReg&quot;, &quot;CD4TCM&quot;, &quot;CD4NaiveT&quot;))
pbmc.14k.eset.log2$celltype &lt;- indx[pbmc.14k.eset.log2$ClusterRes]
</code></pre>
<h2 id="network-generation-via-sjaracne">Network generation via SJARACNe</h2>
<h3 id="generate-sjaracne-input">Generate SJARACNe input</h3>
<p><code>generateSJAracneInput()</code> functions uses the annotated cell type information stored in <code>pData([your_expressionSet])</code> to generate inputs for running SJARACNe from the command line. The function partitions the input expression matrix and performs essential filtering (filter out not expressed genes in a cluster) to ensure a reliable network construction. <code>funcType</code> is required to specify the type of the network. If set <code>funcType="TF"</code>, a reference transcription factor list will be loaded automatically without manual input. However, you do need to define species information for your data using under <code>ref</code>.</p>
<p>This function creates one directory containing the required inputs (filtered expression matrix in .exp format and filtered TF list in .txt format) for each cell type. </p>
<pre><code class="language-R">generateSJARACNeInput(
    input_eset = eset, funcType = &quot;TF&quot;, 
    ref = &quot;hg&quot;,  #human
    wd.src = &quot;SJARACNe&quot;,  #Output directory
    group_tag = &quot;celltype&quot;)
</code></pre>
<h3 id="run-sjaracne-from-the-command-line">Run SJARACNe from the command line</h3>
<p>SJARACNe works as a separate module which was implemented in python and C++, refer to <a href="https://github.com/jyyulab/SJARACNe">SJARACNe</a> github page for installation and basic usage. Please save your working directory before running SJARACNe. </p>
<p>Here we provide an example to print SJARACNe commands for IBM LSF platform.  </p>
<pre><code>indir = ./SJARACNe
cd $indir
for i in $(ls -d */ | cut -f1 -d'/'); do
echo sjaracne lsf -j $indir/config_cwlexec.json -e $indir/${i}/*.exp -g $indir/${i}/tf/*.txt -n 100 -o $indir/${i}/tf -pc 0.01;
done
</code></pre>
<h2 id="identify-cell-type-specific-hidden-drivers-via-minie">Identify cell-type-specific hidden drivers via MINIE</h2>
<p>Identify hidden driver from content-based network is the key step in scMINER to help understand your scRNA-seq data, and provide biological insight. </p>
<h3 id="calculate-activity">Calculate activity</h3>
<p>Activity calculation is the basis of driver estimation in scMINER. To infer driver activity, expression profile of their targets are intergrated via function <code>GetActivityFromSJARACNe</code>. This function takes SJARACNe output path and expression set as input, and return an activity set as well as structured network files if set <code>save_network_files=TRUE</code>. <strong>Please note that this function could only work if you used <code>generateSJARACNeInput</code> to create SJARACNe input directory and files.</strong></p>
<p>Since scRNA-seq data are extremly sparse and noisy, please set <code>activity.method</code> as <code>'unweighted'</code>. </p>
<pre><code class="language-R">acs.14k &lt;- GetActivityFromSJARACNe(
    SJARACNe_output_path =&quot;SJARACNE/&quot;,
    SJARACNe_input_eset = eset,
    activity.method=&quot;unweighted&quot;, # we highly recommend using 'unweighted' as activity calculation method
    activity.norm=TRUE, 
    group_tag = &quot;celltype&quot;, # which group was used to partition expression profiles
    save_network_file=TRUE, # whether or not save network for each group
    save_path=&quot;./networks/&quot;) #default as false, but recommended to be TRUE
</code></pre>
<h3 id="driver-estimation-by-differential-activity-analysis">Driver estimation by differential activity analysis</h3>
<p>The function <code>get.DA</code> was designed to perform differnetial activity analysis from SJARACNe inferred activity matrix. In this function, two-sided student's t-test will be performed to compare mean activity from one cell type V.S. the others. It will return a data frame that includes all TF occurred in original data. Statistics such as t.statistics, p-value, 95%CI, etc. are outputed to help identify hidden drivers. You can save it to file in order to check them manually. </p>
<pre><code class="language-R">DAG_result &lt;- get.DA(input_eset = acs.14k,group_tag = &quot;celltype&quot;)
</code></pre>
<p>We also offer a function called <code>get.Topdrivers</code> to help picking top drivers for each cell type. You can specify <code>n</code> as maximum number of top drivers to pick, and <code>degree_filter</code> to restrict number of targets. </p>
<pre><code class="language-R">TF_list &lt;-get.Topdrivers(DAG_result = DAG_result,
                             celltype = levels(acs.14k$celltype), # ensure cluster order
                             n = 5, degree_filter = c(50,600))
</code></pre>
<p>In scMINER, we provide a handful of visualizations to compare driver activity from different cell type/clusters. Here we demo two basic functions: <code>feature_heatmap</code> and <code>feature_vlnplot</code>. These functions could be used on either expression and activity matrix.</p>
<pre><code class="language-R">feature_heatmap(input_eset = acs.14k, target = TF_list, group_tag = &quot;celltype&quot;,feature = &quot;geneSymbol&quot;,
                width = 6,height = 6, save_plot=TRUE, cluster_rows = FALSE,
                name = &quot;Activity&quot;,plot_name=&quot;./21_TopTFHeatmap.png&quot;)
</code></pre>
<p><center><img src="../plots/4_1_TopTFHeatmap.png" alt="Driver heatmap" width="550"></center></p>
<pre><code class="language-R">#check postive controls
p&lt;-feature_vlnplot(input_eset = acs.14k,feature = &quot;geneSymbol&quot;,target=c(&quot;LEF1&quot;,&quot;TCF7&quot;,&quot;BATF&quot;,&quot;TBX21&quot;,&quot;IRF8&quot;,&quot;SPIB&quot;,&quot;BATF3&quot;,&quot;CEBPA&quot;),
                    ylabel = &quot;Activity&quot;,group_by = &quot;celltype&quot;, ncol=2)
</code></pre>
<p><center><img src="../plots/4_2_Known_MR_vlnplot.png" alt="Driver heatmap" width="600"></center></p>
<p>In order to perform more advanced network analysis utilizing SJARACNe generated cell type specific networks, please infer detailed guidance under <a href="../Advanced analysis/PBMC-14k-network">Advanced analysis</a> tab.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../CD8T/" class="btn btn-neutral float-right" title="KO vs. WT analysis on CD8+ T cells">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../../contact/" class="btn btn-neutral" title="Contact"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
      <p>Apache License (Version 2.0)</p>
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/jyyulab/scMINER/" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../../contact/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../CD8T/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme_extra.js" defer></script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
