<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Read the input data generated by 10x Genomics from a directory — readInput_10x.dir • scMINER</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Read the input data generated by 10x Genomics from a directory — readInput_10x.dir"><meta name="description" content='This function is used to read the gene expression data from a
directory containing three files generated by 10x Genomics: matrix.mtx,
barcodes.tsv and
features.tsv (or genes.tsv). This function can handle these conditions well:
Alternative file names for feature data: features.tsv by CellRanger &amp;gt; 3.0, and genes.tsv by CellRanger &amp;lt; 3.0;
One or more input files are compressed, usually in ".gz" format;
Data with multiple modalities: like the single cell multiome data. In this case, it only retains the data of "Gene Expression".

'><meta property="og:description" content='This function is used to read the gene expression data from a
directory containing three files generated by 10x Genomics: matrix.mtx,
barcodes.tsv and
features.tsv (or genes.tsv). This function can handle these conditions well:
Alternative file names for feature data: features.tsv by CellRanger &amp;gt; 3.0, and genes.tsv by CellRanger &amp;lt; 3.0;
One or more input files are compressed, usually in ".gz" format;
Data with multiple modalities: like the single cell multiome data. In this case, it only retains the data of "Gene Expression".

'></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scMINER</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/quick_tutorial.html">Quick Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="../bookdown/index.html">Full Documentation</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/jyyulab/scMINER"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Read the input data generated by 10x Genomics from a directory</h1>

      <div class="d-none name"><code>readInput_10x.dir.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is used to read the gene expression data from a
directory containing three files generated by 10x Genomics: <strong>matrix.mtx</strong>,
<strong>barcodes.tsv</strong> and
<strong>features.tsv</strong> (or <strong>genes.tsv</strong>). This function can handle these conditions well:</p><ul><li><p>Alternative file names for feature data: <strong>features.tsv</strong> by CellRanger &gt; 3.0, and <strong>genes.tsv</strong> by CellRanger &lt; 3.0;</p></li>
<li><p>One or more input files are compressed, usually in "<strong>.gz</strong>" format;</p></li>
<li><p>Data with multiple modalities: like the single cell multiome data. In this case, it only retains the data of "Gene Expression".</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">readInput_10x.dir</span><span class="op">(</span></span>
<span>  <span class="va">input_dir</span>,</span>
<span>  featureType <span class="op">=</span> <span class="st">"gene_symbol"</span>,</span>
<span>  removeSuffix <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  addPrefix <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-input-dir">input_dir<a class="anchor" aria-label="anchor" href="#arg-input-dir"></a></dt>
<dd><p>Path to the directory containing the 3 files generated by
10x Genomics: <strong>matrix.mtx</strong>, <strong>barcodes.tsv</strong> and <strong>features.tsv</strong> (or
<strong>genes.tsv</strong>)</p></dd>


<dt id="arg-featuretype">featureType<a class="anchor" aria-label="anchor" href="#arg-featuretype"></a></dt>
<dd><p>Character, feature type to use as the gene name of
expression matrix: <code>"gene_symbol"</code> (the default) or <code>"gene_id"</code>.</p></dd>


<dt id="arg-removesuffix">removeSuffix<a class="anchor" aria-label="anchor" href="#arg-removesuffix"></a></dt>
<dd><p>Logical, whether to remove the suffix "<strong>-1</strong>" when
present in all cell barcodes. Default: <code>TRUE</code>.</p></dd>


<dt id="arg-addprefix">addPrefix<a class="anchor" aria-label="anchor" href="#arg-addprefix"></a></dt>
<dd><p>Character or <code>NULL</code>, add a prefix to the cell barcodes, like
Sample ID. It is highly recommended to use a prefix containing letters
and/or numbers only, and not starting with numbers. Default: <code>NULL</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A sparse gene expression matrix of raw UMI counts, genes by cells</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">input_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/demo_inputs/cell_matrix_10x"</span>, package <span class="op">=</span> <span class="st">"scMINER"</span><span class="op">)</span> <span class="co"># path to input data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">input_dir</span>, full.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># you should see three files: matrix.mtx, barcodes.tsv and features.tsv (or genes.tsv)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "barcodes.tsv.gz" "features.tsv.gz" "matrix.mtx.gz"  </span>
<span class="r-in"><span><span class="va">sparseMatrix</span> <span class="op">&lt;-</span> <span class="fu">readInput_10x.dir</span><span class="op">(</span><span class="va">input_dir</span>,</span></span>
<span class="r-in"><span>                                  featureType <span class="op">=</span> <span class="st">"gene_symbol"</span>,</span></span>
<span class="r-in"><span>                                  removeSuffix <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                                  addPrefix <span class="op">=</span> <span class="st">"demoSample"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading 10x Genomcis data from: /private/var/folders/v0/njhqcmrs32xgrjgx2wz8d50r0000gp/T/RtmpDBzcqU/temp_libpathfadd30e717eb/scMINER/extdata/demo_inputs/cell_matrix_10x ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 	Multiple data modalities were found: Gene Expression, Peaks . Only the gene expression data (under "Gene Expression") was kept.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Done! The sparse gene expression matrix has been generated: 500 genes, 100 cells.</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Qingfei Pan, Jiayu Zhou, Liang Ding, Hao Shi, Xinran Dong, Yu Lab @ St Jude Children’s Research Hospital.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

