
#' Read the input data generated by 10x Genomics from a directory
#'
#' @description This function is used to read the gene expression data from a
#' directory containing three files generated by 10x Genomics: **matrix.mtx**,
#' **barcodes.tsv** and
#' **features.tsv** (or **genes.tsv**). This function can handle these conditions well:
#' - Alternative file names for feature data: **features.tsv** by CellRanger > 3.0, and **genes.tsv** by CellRanger < 3.0;
#' - One or more input files are compressed, usually in "**.gz**" format;
#' - Data with multiple modalities: like the single cell multiome data. In this case, it only retains the data of "Gene Expression".
#'
#' @param input_dir Path to the directory containing the 3 files generated by
#'   10x Genomics: **matrix.mtx**, **barcodes.tsv** and **features.tsv** (or
#'   **genes.tsv**)
#' @param featureType Character, feature type to use as the gene name of
#'   expression matrix: `"gene_symbol"` (the default) or `"gene_id"`.
#' @param removeSuffix Logical, whether to remove the suffix "**-1**" when
#'   present in all cell barcodes. Default: `TRUE`.
#' @param addPrefix Character or `NULL`, add a prefix to the cell barcodes, like
#'   Sample ID. It is highly recommended to use a prefix containing letters
#'   and/or numbers only, and not starting with numbers. Default: `NULL`.
#'
#' @return A sparse gene expression matrix of raw UMI counts, genes by cells
#' @export
#'
#' @examples
#' input_dir <- system.file("extdata/demo_inputs/cell_matrix_10x", package = "scMINER") # path to input data
#' list.files(input_dir, full.names = FALSE) # you should see three files: matrix.mtx, barcodes.tsv and features.tsv (or genes.tsv)
#' sparseMatrix <- readInput_10x.dir(input_dir,
#'                                   featureType = "gene_symbol",
#'                                   removeSuffix = TRUE,
#'                                   addPrefix = "demoSample")
readInput_10x.dir <- function(input_dir,
                              featureType = "gene_symbol",
                              removeSuffix = TRUE, addPrefix = NULL)
{
  ## check packages and parameters
  if (!dir.exists(input_dir)) {
    stop("The input directory was not found: ", input_dir, ".")
  }

  cat("Reading 10x Genomcis data from:", input_dir, "...\n")
  f <- list.files(path = input_dir, full.names = F)

  ## matrix
  f_matrix <- f[grepl("matrix.mtx", f)]
  if (length(f_matrix) == 0) {
    stop("The expression matrix file (matrix.mtx[.gz]) was not found in", input_dir, ". Please check and re-try.")
  } else if (length(f_matrix) == 1) {
    d_matrix <- Matrix::readMM(file.path(input_dir, f_matrix))
  } else {
    stop("Multiple expression matrix files (matrix.mtx[.gz]) were found in", input_dir, ". Please check and re-try.")
  }

  ## barcodes
  f_barcode <- f[grepl("barcodes.tsv", f)]
  if (length(f_barcode) == 0) {
    stop("The cell barcode file (barcodes.tsv[.gz]) was not found in", input_dir, ". Please check and re-try.")
  } else if (length(f_barcode) == 1) {
    d_barcode <- read.delim(file.path(input_dir, f_barcode), header = FALSE, stringsAsFactors = FALSE, sep = "\t", quote = "")
  } else {
    stop("Multiple cell barcode files (barcodes.tsv[.gz]) were found in", input_dir, ". Please check and re-try.")
  }

  ## features
  f_feature <- f[grepl("genes.tsv|features.tsv", f)]
  if (length(f_feature) == 0) {
    stop("The gene/feature file (genes.tsv[.gz] or features.tsv[.gz]) was not found in", input_dir, ". Please check and re-try.")
  } else if (length(f_feature) == 1) {
    d_feature <- read.delim(file.path(input_dir, f_feature), header=FALSE, stringsAsFactors = FALSE, sep = "\t", quote = "")
  } else {
    stop("Multiple gene/feature files (genes.tsv[.gz] or features.tsv[.gz]) were found in", input_dir, ". Please check and re-try.")
  }

  if (ncol(d_feature) >= 3) {
    l_modality <- unique(d_feature[,3])
    if ("Gene Expression" %in% l_modality) {
      if (length(l_modality) > 1) {
        cat('\tMultiple data modalities were found:', paste0(l_modality, collapse = ", "), '. Only the gene expression data (under "Gene Expression") was kept.\n')
        d_matrix <- d_matrix[d_feature[,3] %in% c("Gene Expression"),]
        d_feature <- d_feature[d_feature[,3] %in% c("Gene Expression"),]
      }
    } else {
      stop('The gene expression data (under "Gene Expression") was not found. Please check and re-try.')
    }
  }

  geneNames <- c()
  if (featureType == "gene_symbol") {
    geneNames <- base::make.unique(d_feature[,2], sep = "_")
  } else if (featureType == "gene_id") {
    geneNames <- base::make.unique(d_feature[,1], sep = "_")
  } else {
    stop('The featureType was not reconized. Please specify: [gene_symbol | gene_id].')
  }

  cellNames <- d_barcode[,1]
  if (removeSuffix == TRUE) {
    if_matched <- grepl("-1$", cellNames)
    if (all(if_matched)) {
      cellNames <- gsub("-1$", "", cellNames)
    } else {
      cat('\tSuffix removal was specified but skipped, since some barcodes do not carry "-1" suffix.\n')
    }
  }

  if (is.null(addPrefix) == FALSE) {
    if (grepl("[^A-Za-z0-9]|^[0-9]", addPrefix)) {
      warning('\tIt is highly recommended to use a prefix containing letters and/or numbers only, and not starting with numbers.')
    }
    cellNames <- paste0(addPrefix, "_", cellNames)
  }

  dimnames(d_matrix)[[1]] <- geneNames
  dimnames(d_matrix)[[2]] <- cellNames
  cat('Done! The sparse gene expression matrix has been generated:', length(geneNames), 'genes,', length(cellNames), 'cells.\n')

  return(d_matrix)
}


#' Read the input data generated by 10x Genomics from the HDF5 file
#'
#' @description This function is used to read the gene expression data from the
#' HDF5 file generated by CellRanger pipeline of 10x Genomics. This function can
#' automatically distinguish the data of different modalities (e.g. expression
#' data, ATAC data) and retains the gene expression data only. The `**hdf5r**`
#' package is needed to use this function.
#'
#' @param h5_file H5 file generated by CellRanger pipeline of 10x Genomics
#' @param featureType Character, feature type to use as the gene name of
#'   expression matrix: `"gene_symbol"` (the default) or `"gene_id"`.
#' @param removeSuffix Logical, whether to remove the suffix "**-1**" when
#'   present in all cell barcodes. Default: `TRUE`.
#' @param addPrefix Character or `NULL`, add a prefix to the cell barcodes, like
#'   Sample ID. It is highly recommended to use a prefix containing letters
#'   and/or numbers only, and not starting with numbers. Default: `NULL`.
#'
#' @return A sparse gene expression matrix of raw UMI counts, genes by cells
#' @export
#'
#' @examples
#' h5_file <- system.file("extdata/demo_inputs/hdf5_10x/demoData3.h5", package = "scMINER") # path to hdf5 file
#' sparseMatrix <- readInput_10x.h5(h5_file,
#'                                  featureType = "gene_symbol",
#'                                  removeSuffix = TRUE,
#'                                  addPrefix = "demoSample")
readInput_10x.h5 <- function(h5_file,
                             featureType = "gene_symbol",
                             removeSuffix = TRUE,
                             addPrefix = NULL)
{
  ## check packages and parameters
  #
  avail_hdf5r <- require(package = "hdf5r", quietly = TRUE)
  if (avail_hdf5r == FALSE) {
    stop('The package, hdf5r, is required but not installed. Please run install.packages("hdf5r") to install it and re-try!')
  }
  #
  if (featureType == "gene_symbol") {
    featureType <- "name"
  } else if (featureType == "gene_id") {
    featureType <- "id"
  } else {
    stop('The featureType was not reconized. Please specify: [gene_symbol | gene_id].')
  }

  ## read hdf5 file
  if (!file.exists(h5_file)) {
    stop("Input HDF5 file was not found:", h5_file, ".")
  }
  cat('Reading 10x Genomics data from:', h5_file, '...\n')
  f_h5 <- hdf5r::H5File$new(filename = h5_file, mode = "r")

  ## check hdf5 format
  cat("\tChecking HDF5 file format ...\n")
  if (all(f_h5$exists("matrix/indices"), f_h5$exists("matrix/indptr"), f_h5$exists("matrix/data"), f_h5$exists("matrix/shape"), f_h5$exists("matrix/barcodes"), f_h5$exists("matrix/features/feature_type"), f_h5$exists(paste0("matrix/features/", featureType)))) {
    # check the data modality
    l_modality <- unique(f_h5[["matrix/features/feature_type"]][])
    if ("Gene Expression" %in% l_modality) {
      if (length(l_modality) > 1) {
        cat('\tMultiple data modalities were found:', paste0(l_modality, collapse = ", "), '. Only the gene expression data (under "Gene Expression") was kept.\n')
      }
      cat('\tFormat check passed!\n')
    } else {
      stop('The gene expression data (under "Gene Expression") was not found. Please check and re-try.')
    }
  } else {
    stop('\tFormat check failed.\nThese 7 datasets are needed: matrix/data, matrix/barcodes, ', paste0("matrix/features/", featureType), ', matrix/features/feature_type, matrix/indices, matrix/indptr and matrix/shape.')
  }

  ## matrix
  d_matrix <- Matrix::sparseMatrix(i = f_h5[["matrix/indices"]][] + 1,
                                   p = f_h5[["matrix/indptr"]][],
                                   x = as.numeric(f_h5[["matrix/data"]][]),
                                   dims = f_h5[["matrix/shape"]][],
                                   repr = "T")
  d_matrix <- d_matrix[l_modality %in% c("Gene Expression"),]

  ## barcodes
  d_barcode <- f_h5[["matrix/barcodes"]][]
  if (removeSuffix == TRUE) {
    if_matched <- grepl("-1$", d_barcode)
    if (all(if_matched)) {
      d_barcode <- gsub("-1$", "", d_barcode)
    } else {
      cat('Suffix removal was specified but skipped, since some barcodes do not carry "-1" suffix.\n')
    }
  }

  if (is.null(addPrefix) == FALSE) {
    if (grepl("[^A-Za-z0-9]|^[0-9]", addPrefix)) {
      warning('It is highly recommended to use a prefix containing letters and/or numbers only, and not starting with numbers.')
    }
    d_barcode <- paste0(addPrefix, "_", d_barcode)
  }

  ## features
  d_feature <- f_h5[[paste0("matrix/features/", featureType)]][]
  d_feature <- base::make.unique(d_feature)
  d_feature <- d_feature[l_modality %in% c("Gene Expression")]

  dimnames(d_matrix)[[1]] <- d_feature
  dimnames(d_matrix)[[2]] <- d_barcode
  cat('Done! The sparse gene expression matrix has been generated:', length(d_feature), 'genes,', length(d_barcode), 'cells.\n')

  return(d_matrix)
}


#' Read the h5ad file
#'
#' @description This function is used to read the h5ad file, a popular file
#' format for storing and sharing single-cell RNA sequencing data. The
#' `**anndata**` package is needed to use this function.
#'
#' @param h5ad_file H5ad file of sc/snRNA-seq data
#' @param removeSuffix Logical, whether to remove the suffix "**-1**" when
#'   present in all cell barcodes. Default: `TRUE`.
#' @param addPrefix Character or `NULL`, add a prefix to the cell barcodes, like
#'   Sample ID. It is highly recommended to use a prefix containing letters
#'   and/or numbers only, and not starting with numbers. Default: `NULL`.
#'
#' @return A AnnData object containing `"X"` (a observations x variables data
#'   matrix), `"obs"` (data frame of observations), `"var"` (data frame of
#'   variables) and more. For more details, please check out
#'   <https://anndata.readthedocs.io/en/latest/generated/anndata.AnnData.html>.
#' @export
#'
#' @examples
#' h5ad_file <- system.file("extdata/demo_inputs/h5ad_file/demoData4.h5ad", package = "scMINER") # path to h5ad file
#' sparseMatrix <- readInput_h5ad(h5ad_file,
#'                                removeSuffix = FALSE,
#'                                addPrefix = "demoSample")
readInput_h5ad <- function(h5ad_file,
                           removeSuffix = FALSE, addPrefix = NULL)
{
  ## check packages and parameters
  #
  avail_anndata <- require(package = "anndata", quietly = TRUE)
  if (!avail_anndata) {
    stop('The package, anndata, is required but not installed. Please run install.packages("anndata") to install it and re-try!')
  }

  ## read h5ad file
  if (!file.exists(h5ad_file)) {
    stop("Input h5ad file was not found: ", h5ad_file, ".")
  }

  cat('Reading h5ad file:', h5ad_file, '...\n')
  h5ad_obj <- anndata::read_h5ad(h5ad_file, backed = 'r+')

  if (removeSuffix == TRUE) {
    if_matched <- grepl("-1$", row.names(h5ad_obj))
    if (all(if_matched)) {
      row.names(h5ad_obj) <- gsub("-1$", "", row.names(h5ad_obj))
    } else {
      cat('\tSuffix removal was specified but skipped, since some barcodes do not carry "-1" suffix.\n')
    }
  }

  if (is.null(addPrefix) == FALSE) {
    if (grepl("[^A-Za-z0-9]|^[0-9]", addPrefix)) {
      warning('\tIt is highly recommended to use a prefix containing letters and/or numbers only, and not starting with numbers.')
    }
    row.names(h5ad_obj) <- paste0(addPrefix, "_", row.names(h5ad_obj))
  }

  cat('Done! The sparse gene expression matrix has been generated:', dim(h5ad_obj)[2], 'genes,', dim(h5ad_obj)[1], 'cells.\n')

  return(h5ad_obj)
}


#' Read the table format file
#'
#' @description This function is used to read data from a table-format file. The
#' user needs to specify the format of the table using the parameter
#' **`is.geneBYcell`**:
#' - `TRUE` (the default): the rows will be treated as genes, while the columns will be treated as cells;
#' - `FALSE`: the rows will be treated as cells, while the columns will be treated as genes.
#'
#' @param table_file The table format file (e.g. **txt**, **tsv**, **csv**, and
#'   others) which the data are to be read from.
#' @param sep String, The field separator character. Default: `"\t"`.
#' @param is.geneBYcell Logical, whether the table is organized in **gene (row)
#'   by cell (column)** format. If `FALSE`, the rows will be treated as cells.
#'   Default: `TRUE`.
#' @param removeSuffix Logical, whether to remove the suffix "**-1**" when
#'   present in all cell barcodes. Default: `FALSE`.
#' @param addPrefix Character or `NULL`, add a prefix to the cell barcodes, like
#'   Sample ID. It is highly recommended to use a prefix containing letters
#'   and/or numbers only, and not starting with numbers. Default: `NULL`.
#'
#' @return A sparse gene expression matrix, genes by cells
#' @export
#'
#' @examples
#' table_file <- system.file("extdata/demo_inputs/table_file/demoData2.txt.gz", package = "scMINER") # path to text-table file
#' sparseMatrix <- readInput_table(table_file,
#'                                 sep = "\t",
#'                                 is.geneBYcell = TRUE,
#'                                 removeSuffix = FALSE,
#'                                 addPrefix = "demoSample")
readInput_table <- function(table_file, sep = "\t", is.geneBYcell = TRUE,
                            removeSuffix = FALSE, addPrefix = NULL)
{
  if (!file.exists(table_file)) (
    stop('Input file was not found: ', table_file, '.')
  )

  cat('Reading table file:', table_file, '...\n')
  d_matrix <- read.delim(table_file, sep = sep, header = TRUE, row.names = 1, stringsAsFactors = FALSE, quote = "", check.names = FALSE)

  if (is.geneBYcell == FALSE) {
    d_matrix <- t(d_matrix)
  }

  if (removeSuffix == TRUE) {
    if_matched <- grepl("-1$", colnames(d_matrix))
    if (all(if_matched)) {
      colnames(d_matrix) <- gsub("-1$", "", colnames(d_matrix))
    } else {
      cat('\tSuffix removal was specified but skipped, since some barcodes do not carry "-1" suffix.\n')
    }
  }

  if (is.null(addPrefix) == FALSE) {
    if (grepl("[^A-Za-z0-9]|^[0-9]", addPrefix)) {
      warning('\tIt is highly recommended to use a prefix containing letters and/or numbers only, and not starting with numbers.')
    }
    colnames(d_matrix) <- paste0(addPrefix, "_", colnames(d_matrix))
  }

  d_matrix <- Matrix::Matrix(as.matrix(d_matrix), sparse = TRUE)
  cat('Done! The sparse gene expression matrix has been generated:', dim(d_matrix)[1], 'genes,', dim(d_matrix)[2], 'cells.\n')

  return(d_matrix)
}



#' Create a project space for scMINER analysis
#'
#' @description This function is used to create a folder of the specified
#' project names in the specified project directory as the project space to run
#' scMINER analysis. It also creastes 4 subfolers inside of it: "`DATA`",
#' "`MICA`", "`SJARACNe`" and "`PLOT`".
#'
#' @param project_dir The directory to create the project space in
#' @param project_name The name of the project, will be used as the name of the
#'   folder
#' @param do.unlink Logical, whether to remove the files and/or folders inside
#'   of pre-existing project space. Default: `FALSE`.
#'
#' @return It creates a folder of project name and 4 subfolders in the project
#'   directory, and returns the path of project space.
#' @export
#'
#' @examples
#' scminer_dir <- createProjectSpace(project_dir = "path-to-a-folder",
#'                                   project_name = "PBMC14k")
createProjectSpace <- function(project_dir,
                               project_name,
                               do.unlink = FALSE)
{
  ## check parameters
  scminer_dir <- sprintf('%s/%s', project_dir, project_name)

  if (dir.exists(scminer_dir) == TRUE) {
    check_dir <- c(list.files(scminer_dir, recursive = T), list.dirs(scminer_dir, recursive = T))
    if (length(check_dir) > 0) {
      if (do.unlink == FALSE) {
        stop('A folder of the specified project name was found in the specified project directory and it is not empty: ', scminer_dir, '.\nPlease set do.unlink to TRUE to remove the files and subfolder in it, or use another project name.')
      } else {
        base::unlink(scminer_dir, recursive = TRUE)
        dir.create(scminer_dir, recursive = TRUE)
        cat('A folder of the specified project name was found and has been cleaned up in the specified project directory\n')
      }
    } else {
      cat('A folder of the specified project name was found in the specified project directory It can be used as the project space since it is empty.\n')
    }
  } else {
    dir.create(scminer_dir, recursive = TRUE)
  }

  scminer_dir.data <- paste0(scminer_dir, "/DATA")
  if (dir.exists(scminer_dir.data) == FALSE) {dir.create(scminer_dir.data)}

  scminer_dir.mica <- paste0(scminer_dir, "/MICA")
  if (dir.exists(scminer_dir.mica) == FALSE) {dir.create(scminer_dir.mica)}

  scminer_dir.sjaracne <- paste0(scminer_dir, "/SJARACNe")
  if (dir.exists(scminer_dir.sjaracne) == FALSE) {dir.create(scminer_dir.sjaracne)}

  scminer_dir.plot <- paste0(scminer_dir, "/PLOT")
  if (dir.exists(scminer_dir.plot) == FALSE) {dir.create(scminer_dir.plot)}

  cat("The project space has been successfully created:", scminer_dir, "!\n")

  return(scminer_dir)
}
